# Preprocessing
## Declare libraries
```{r}
suppressPackageStartupMessages({
  library(Peptides)
  library(data.table)
  library(seqinr)
  library(tidyverse)
  library(circlize)
  library(patchwork)
  library(readxl)
  library(ComplexHeatmap)
  library(RColorBrewer)
  library(stringi)
})
```

## Define directories, files and colors
```{r}
# Work directories
work_dir <- paste0("<work_directory>")
raw_dir <- file.path(work_dir, "raw")
processed_dir <- file.path(work_dir, "processed")
plot_dir <- file.path(processed_dir, "plots")

# For running predictions yourself specify
## Directory used for NetMHCpan binding predictions input
net_input <- file.path(processed_dir, "netmhcpan_input")
## Directory containg binding predictions for detected peptides
net_output <- file.path(processed_dir, "netmhcpan_output_annotatedalleles")
# Or use data previously generated (from github)
net_output_rds <- file.path(raw_dir, "netmhcpan.RDS")

# List of 7,264 ncORFs from doi: 10.1038/s41587-022-01369-0
orf_file <- file.path(raw_dir, "ncorf_list.xlsx")

# Peptide files and file mapping peptides to proteins/ncORFs from PeptideAtlas
peptide_file <- file.path(raw_dir, "peptide_sample_msrun_counts.tsv")
## Either needs mapping file
mapping_file <- file.path(raw_dir, "peptide_mapping.tsv")
## Or filtered peptides file
filtered_peptides_file <- file.path(raw_dir, "filtered_peptides_file")

# Protein sequences
## Either needs file containing the sequences of proteins/ncORFs from PeptideAtlas
seq_file <- file.path(raw_dir, "Homo_sapiens.fasta")
## Or prot_seq_df file
can_nonc_seq_file <- file.path(raw_dir, "can_nonc_seq.RDS")

# List of canonical CDSs from PeptideAtlas
canonical_file <- file.path(raw_dir, "Core20k.txt")



# Experiment statistics for the HLA and nonHLA builds from PeptideAtlas
atlas_stats_file <- file.path(raw_dir, "HLA2023-09_experiment_summary.xlsx")
atlas_stats_nonhla_file <- 
  file.path(raw_dir, "Non-HLA2023-09_experiment_summary.tsv")

# List of genes associated with cancer from the Cancer Gene Census
cancer_census_file <- 
  file.path(raw_dir, "Census_allThu_Jan_4_14_08_58_2024.csv")

# Supplementary tables about detected peptides and ncORFs
table_s2_file <- file.path(raw_dir, "060924_Supp_Table_S2.xlsx")
table_s3_file <- file.path(raw_dir, "060924_Supp_Table_S3.xlsx")
table_s4_file <- file.path(raw_dir, "060924_Supp_Table_S4.xlsx")
table_s5_file <- file.path(raw_dir, "060924_Supp_Table_S5.xlsx")

# Suplementary table S6 containing MS-runs and their annotation
annotation_file <- file.path(raw_dir, "060924_Supp_Table_S6.xlsx")

# GTF file
gtf_file <- file.path("Homo_sapiens.GRCh38.102.gtf")

# GTEX FPKM mean expression data (excluding testis)
gtex_fpkm_file <- file.path(raw_dir, "GTEX_FPKMmean_expression.txt")


blue_col <- c("#6C88A5", "#7B95B3", "#8BA3C1", "#9AB1CF", "#AAC0DD", "#B9CEE9", "#C9DBF7", "#D8E9FF", "#E7F6FF", "#F5FFFF")
grey_col <- c("#7F7F7F", "#8E8E8E", "#9D9D9D", "#ABABAB", "#BABABA", "#C9C9C9", "#D8D8D8", "#E6E6E6", "#F5F5F5", "#FFFFFF")
pink_col <- c("#EFA891", "#F2B09F", "#F5C8AD", "#F7E1BC", "#FAF9CA", "#FCFAD8", "#FFFBF6", "#FFFCFF", "#FFFDFF", "#FFFFFE")
green_col <- c("#7D905D", "#8D9E6C", "#9DA97B", "#ADB58A", "#BCC19A", "#CBDCA9", "#DAE8B8", "#EAF3C8", "#F9FFD7", "#FFFFE6")
purple_col <- c("#9F85A1", "#AD92AF", "#BCAFC1", "#CBBDD0", "#D9CAE0", "#E8D7EF", "#F6E5FF", "#FFF2FF", "#FFFFFE", "#FFFFFF")

red_col <- c("#B10027")

# Standard theme set for plots
black_theme <- theme_classic() + 
  theme(text = element_text(colour = "black"), 
        axis.title = element_text(size = 7, colour = "black"),
        axis.text = element_text(size = 7, colour = "black"),
        axis.ticks = element_line(color = "black"),
        legend.text = element_text(size = 7, colour = "black"),
        legend.title = element_text(size = 7, colour = "black")) 
  

remove_y <- theme(
  axis.text.y = element_blank(),
  axis.ticks.y = element_blank(),
  axis.title.y = element_blank()
)

remove_x <- theme(
  axis.text.x = element_blank(),
  axis.ticks.x = element_blank(),
  axis.title.x = element_blank()
)
```

## Declare functions

```{r}
get_psites <- function(p_sites, p_chr, p_start, p_end) {
  # Extract P-sites add specified coordinates and ungroup grouped p-sites
  sel_psites <- p_sites %>% filter(chr == p_chr,start > p_start - 100,
                                  end < p_end + 100) %>% 
  mutate(diff = end - start) %>% 
  uncount(diff) %>% group_by(chr,start,end) %>% 
  mutate(dup = row_number(), start = start + dup - 1, end = start + 1) %>% 
  dplyr::count(chr,start, end, wt = n) %>% ungroup() %>% 
  mutate(group = end%%3,
         group = factor(group, levels = c(0:2))) %>% 
    dplyr::rename(psite_x = end) %>% select(-start)
  return(sel_psites)
}

write_net <- function(sample_num, net_data) {
  # Write peptides to net_input folder for netMHCpan predictions
  sample_name <- paste0("/netmhcpan_input", sample_num, ".txt")
  net_data <- net_data %>% dplyr::select(peptide) %>% distinct()
  write.table(net_data, paste0(net_input, sample_name), row.names = F, 
              col.names = F, quote = F)
}

read_netmhcpan <- function(net_path) {
  # Read netMHCpan predictions to dataframe
  net_data <- read.delim(net_path) %>% distinct()
  rank_data <- net_data[2:nrow(net_data),
                        c(2, which(grepl("EL_Rank", net_data[1,])))]
  colnames(rank_data) <- c("peptide", "rank")
  rank_data$hla_type <- colnames(net_data)[grepl("HLA", colnames(net_data))][1] %>% 
    str_replace_all("HLA|[.]", "")
  rank_data <- rank_data %>% mutate(rank = as.numeric(rank))
  return(rank_data)
}

extract_kmers <- function(string, k) {
  # Extract all kmers of length k from string
  num_kmers <- nchar(string) - k + 1
  if (num_kmers < 1) {
    return(character(0))
  }
  kmers <- sapply(1:num_kmers, function(i) substring(string, i, i + k - 1))
  return(kmers)
}

create_dis_plot <- function(plot_data, x_name, plot_colors, x_breaks = waiver(),
                            y_breaks = waiver(), y_name = "Density") {
  # Create density plot using ggplot() as input                          
  plot_data + geom_density(alpha=.4, adjust = 1.5, linewidth = 0.5) + 
    black_theme +
    scale_x_continuous(expand = c(0, 0), breaks = x_breaks, name = x_name) +
    scale_y_continuous(expand = c(0, 0), breaks = y_breaks, name = y_name) +
    scale_color_manual(values = plot_colors) + 
    theme(legend.position = "none")
}

get_stats <- function(df) {
  # Get protein characteristics
  df <- df %>% mutate(len = nchar(protein),
                      hydro = hydrophobicity(protein, "KyteDoolittle"),
                      pi = pI(protein),
                      chrg = charge(protein))
  return(df)
}


perform_wtest <- function(df, cor_num, raw = F) {
  # Perform wilcox test and optionally process result to scientific format
  w_test <- wilcox.test(y_val ~ detected, 
                        data = df %>% mutate(detected = factor(
                          detected, levels = c("Detected", "Undetected"))))
  p_val <- w_test$p.value * cor_num
  if (raw) {
    return(p_val)
  }
  if (p_val >= 0.05) {
    p_val = "p = n.s."
  } else if (p_val >= 0.010) {
    p_val = paste0("p = ", round(p_val, 3))
  } else if (p_val == 0){
    p_val = "p = 0"
  } else {
    p_val = paste0("p = ", format(p_val, scientific = T, digits = 4) %>% 
                     str_replace("e", "\u00D710")) 
  }
  return(p_val)
}

get_atlas_stats <- function(stat_df) {
  # Read PeptideAtlas statistics about the number of detected peptides and proteins
  colnames(stat_df) <- colnames(stat_df) %>% str_replace_all(" |'", ".") %>% 
    str_replace_all("X[.]Spectra", "%Spectra")
  atlas_stats <- stat_df %>% 
    mutate(cum_pep = cumsum(Added.Peptides),
           spectra_searched = cumsum(Spectra.ID.d),
           cum_pep_dis = Distinct.Peptides,
           cum_pro = cumsum(Added.Canonical.Proteins),
           cum_pro_dis = Distinct.Canonical.Proteins,
           count = 2) %>% 
    uncount(count) %>% 
    add_row(spectra_searched = 0, .before = 1) %>% 
    mutate(cum_pep = lead(cum_pep),
           cum_pep_dis = lead(cum_pep_dis),
           cum_pro = lead(cum_pro),
           cum_pro_dis = lead(cum_pro_dis)) %>% 
    filter(!is.na(cum_pep)) %>% 
    select(spectra_searched, cum_pep, cum_pep_dis, cum_pro, cum_pro_dis) %>% 
    pivot_longer(c(c(cum_pep, cum_pep_dis), c(cum_pro, cum_pro_dis))) %>% 
    mutate(name = factor(name, levels = c("cum_pep", "cum_pep_dis", "cum_pro", "cum_pro_dis")))
  return(atlas_stats)
}

create_violin <- function(violin_data, y_label, p_val, y_lim, 
                          n_colors = c(1:8), log = F, pseudo_log = F, 
                          x_label = "", y_breaks = waiver()) {
  # Create violin plot
  counts <- violin_data %>% dplyr::count(detected) %>% 
    mutate(n = format(n, big.mark = ",")) %>% deframe()
  p_label <- grobTree(textGrob(p_val, x = 0.5,  y = 1, just = c("center", "top"),
      gp = gpar(col = "black", fontsize = 7)))
  v_colors <- c(blue_col[c(1,4)], green_col[c(1,4)], purple_col[c(1,4)], 
                pink_col[c(1,4)])[n_colors]
  v_plot <- ggplot(violin_data, aes(order_ids, y = y_val, fill = fill_group)) +
    geom_violin(trim = T, show.legend = F) + 
    geom_boxplot(fill = "transparent", outlier.shape = NA) + black_theme +
    scale_fill_manual(values = v_colors) +
    scale_x_discrete(name = ifelse(x_label == "", element_blank(), x_label), 
                     labels = c(paste0("Detected\n(n = ", 
                                       counts[["Detected"]], ")"),
                                paste0("Undetected\n(n = ", 
                                       counts[["Undetected"]], ")"))) +
    annotation_custom(p_label)
  if (log) {
    v_plot <- v_plot + scale_y_log10(name = y_label, limits = y_lim, breaks = y_breaks)
  } else if (pseudo_log) {
    v_plot <- v_plot + scale_y_continuous(
      trans = scales::pseudo_log_trans(base = 10),
      name = y_label, limits = y_lim, breaks = y_breaks)
  } else {
    v_plot <- v_plot + scale_y_continuous(name = y_label, limits = y_lim, breaks = y_breaks) 
  }
  return(v_plot)
}

make_biotype_levels <- function(df) {
  # Transforms the orf_biotype column in df to a factor
  df %>% 
    mutate(orf_biotype = str_replace(orf_biotype, "Processed transcript ORFs", 
                                     "Processed\ntranscript ORFs"),
           orf_biotype = factor(orf_biotype, 
                              levels = c("uORFs", "uoORFs", "intORFs", "doORFs", 
                                         "dORFs", "lncRNA ORFs", 
                                         "Processed\ntranscript ORFs")),)
}


perform_fisher_test <- function(df) {
  # Function to perform Fisher's test
  df$p_val <- fisher.test(df %>% select(-fc))$p.value
  return(df)
}

fisher_test_ligand_atlas <- function(a1, a2, b1, b2) {
  # Function to perform Fisher's test using four separate inputs
  f_test <- fisher.test(data.frame(a = c(a1,a2), b = c(b1,b2)))
  return(f_test$p.value)
}

process_data <- function(data, filter_condition, offset_col, numerator, denominator) {
  # Compute fold change using the fiven conditions
  data %>%
    filter({{filter_condition}}) %>%
    ungroup() %>%
    count(!!sym(offset_col)) %>%
    pivot_wider(names_from = !!sym(offset_col), values_from = n, values_fill = 0) %>%
    mutate(fc = !!sym(numerator) / !!sym(denominator)) 
}

create_histogram <- function(data, x_var, fill_color, y_label = NULL, 
                             x_label = NULL, y_limits, y_breaks = NULL, 
                             y_labels = NULL) {
  # Create a histogram plot
  ggplot(data, aes(x = {{x_var}})) +
    geom_histogram(binwidth = 1, fill = fill_color, color = "white", 
                   linewidth = 0.2) +
    black_theme +
    scale_x_continuous(expand = c(0, 0), name = x_label) +
    scale_y_continuous(expand = c(0, 0), name = y_label,
                       limits = y_limits, breaks = y_breaks, 
                       labels = y_labels) +
    theme(axis.ticks = if(is.null(y_label) && is.null(x_label)) element_blank() else element_line(),
          axis.ticks.x = if(is.null(x_label)) element_blank() else element_line(),
          axis.text.x = if(is.null(x_label)) element_blank() else element_text(),
          axis.ticks.y = if(is.null(y_label)) element_blank() else element_line())
}
```

## Load initial files and data
### Load files
```{r}
# Load data
orf_sequences <- read_excel(orf_file, sheet = 3)[
  , c("orf_name", "orf_biotype", "orf_sequence", "gene_name", "gene_id", "transcript", "strand", "chrm", "starts", "ends")] %>% 
  bind_rows(read_excel(orf_file, sheet = 4)[
    , c("orf_name", "orf_biotype", "orf_sequence", "gene_name", "gene_id", "transcript", "strand", "chrm", "starts", "ends")]) %>% 
  mutate(orf_sequence = str_replace(orf_sequence, "[*]", ""),
         orf_biotype = ifelse(orf_biotype == "processed_transcript", "Processed transcript ORF", 
                              orf_biotype),
         orf_biotype = ifelse(orf_biotype == "lncRNA", "lncRNA ORF", orf_biotype),
         orf_biotype = str_replace(orf_biotype, "ORF", "ORFs"),
         protein_accession = paste0("CONTRIB_GENCODE_", orf_name)) %>% 
  dplyr::rename(protein = orf_sequence)

gtex_data <- read.csv(gtex_fpkm_file)

can_id <- read_tsv(canonical_file, col_names = "protein_accession")

annotations <- read_excel(annotation_file) %>% 
  rename(dataset_id = dataset, ms_name = ms_run_name, hla_class = HLA_class) %>% 
  mutate(is_cancer = case_when(grepl("TUMORAL", disease_state) ~ "Cancer", 
                               TRUE ~ "Non-cancer"),
         is_cellline = ifelse(str_detect(sample_category, 
                                         regex("Cell Line", ignore_case = T)), 
                              "cell line", "non-cell line"),
         sample_type = paste0(is_cancer, "_", is_cellline))

peptide <- fread(peptide_file)
colnames(peptide) <- c("peptide", "sam_id", "count", "dataset_id", 
                       "sample_tag", "ms_name")

table_s2 <- read_excel(table_s2_file) %>% 
  dplyr::rename(identifier = "PeptideAtlas identifier",
                orf_name = "Ribo-Seq_ORF")
table_s3 <- read_excel(table_s3_file) %>% 
  dplyr::rename(identifier = "PeptideAtlas.identifier",
                orf_name = "Ribo-Seq_ORF",
                initial_tier = "Initial.tier.for.level.of.evidence.prior.to.manual.inspection",
                final_tier = "Final.tier.for.level.of.evidence")
table_s4 <- read_excel(table_s4_file) %>% 
  dplyr::rename(identifier = "PeptideAtlas.identifier",
                orf_name = "Ribo-Seq_ORF")
table_s5 <- read_excel(table_s5_file) %>% 
  dplyr::rename(identifier = "PeptideAtlas.identifier",
                orf_name = "Ribo-Seq_ORF",
                initial_tier = "Initial.tier.for.level.of.evidence.prior.to.manual.inspection",
                final_tier = "Final.tier.for.level.of.evidence") %>% rename_all(~ make.names(.))
```

### Load protein sequences
#### Either use PeptideAtlas fasta file
```{r}
prot_seq_df <- read.fasta(seq_file, seqtype = "AA", as.string = T) %>% 
  enframe() %>% dplyr::rename(protein_accession = name, protein = value) %>% 
  mutate(protein = as.character(protein))

can_nonc_seq <- prot_seq_df %>% semi_join(can_id, "protein_accession") %>% 
  bind_rows(prot_seq_df %>% 
              filter(str_starts(protein_accession, "CONTRIB_GENCODE_c")))
```

#### Or use pre-generated file (from github)
```{r}
can_nonc_seq <- readRDS(can_nonc_seq_file)
```

### Filter canonical and non-canonical peptides
#### Use mapping file from PeptideAtlas
```{r}
mapping <- fread(mapping_file)
colnames(mapping) <- c("peptide_accession", "peptide", "protein_accession", "start_loc", "end_loc", "prev_aa", "next_aa")

can_peptides <- mapping %>% inner_join(can_id, "protein_accession") %>% distinct(peptide)

mapping <- mapping %>% semi_join(peptide, "peptide")
setDT(mapping)
mapping <- mapping[!str_starts(protein_accession, "DECOY|CONTRIB_smORFs_Cui|CONTRIB_sORFs|CONTRIB_Fedor|CONTRIB_Bazz|CONTRIB_HLA")] 

#Filter noncanonical peptides
pep_nonc_long <- mapping[str_starts(protein_accession, "CONTRIB_GENCODE")] %>% 
  distinct(peptide) %>% inner_join(mapping, "peptide") %>% 
  filter(!str_starts(protein_accession, 
                     "DECOY|CONTRIB_smORFs_Cui|CONTRIB_sORFs|CONTRIB_Fedor|CONTRIB_Bazz|CONTRIB_HLA|CONTRIB_GENCODE_nearcognate")) %>% 
  anti_join(can_peptides, "peptide") %>% group_by(peptide) %>% mutate(n_pep = n()) %>% 
  filter(n_pep <= 10) %>%
  inner_join(orf_sequences, "protein_accession") %>% 
  mutate(pep_len = nchar(peptide)) %>% filter(pep_len >= 8) %>% ungroup()

pep_can <- mapping %>% 
  semi_join(can_peptides, "peptide") %>% group_by(peptide) %>% 
  mutate(n_pep = n(), pep_len = nchar(peptide)) %>% 
  filter(n_pep <= 30, pep_len >= 8)

```
#### Or use earlier generated data (from Github)
```{r}
load(filtered_peptides_file)
```
### Prepare initial data frames for analysis
```{r}

pep_nonc <- pep_nonc_long %>% 
  inner_join(table_s4 %>% dplyr::rename(other_mappings = "Other.mappings"),
    c("peptide" =  "sequence", "protein_accession" = "identifier")) %>% ungroup()

# Filter canonical peptides
setDT(mapping)
setDT(can_peptides)



seq_stats <- can_nonc_seq %>% filter(!str_detect(protein, "U")) %>% 
  mutate(pro_cat = ifelse(str_starts(protein_accession, "CONTRIB_GENCODE_"),
                          "Non-canonical", "Canonical"),
         len = nchar(protein)) %>% 
  filter(len >= 16) %>% 
  left_join(
    bind_rows(pep_can %>% ungroup() %>% distinct(protein_accession) %>% 
                semi_join(can_id, "protein_accession"),
              pep_nonc_long %>% distinct(protein_accession)) %>% 
      mutate(detected = "Detected"), "protein_accession") %>% 
  left_join(orf_sequences %>%  distinct(protein_accession, orf_biotype), 
            "protein_accession") %>% 
  replace_na(list(detected = "Undetected", orf_biotype = "CDS"))


# Combine peptides and annotation, filter for HLA-I
peptide_hla <- peptide %>% 
  inner_join(annotations, c("ms_name", "dataset_id"))

pep_can_nonc <- peptide %>% distinct(peptide) %>% 
  left_join(bind_rows(pep_can %>% distinct(peptide) %>%
                        mutate(pep_type = "canonical"),
                      pep_nonc %>% distinct(peptide) %>% 
                        mutate(pep_type = "noncanonical")),
            "peptide") %>% 
  mutate(pep_type = ifelse(is.na(pep_type), "other", pep_type))

setDT(pep_can_nonc)
setDT(peptide_hla)
peptide_hla <- peptide_hla[pep_can_nonc, on = 'peptide']

hla_df <- annotations %>% 
  select(dataset_id, ms_name, sample_type, is_cancer, is_cellline, hla_class, 
         hla_typing) %>% 
  filter(hla_class == "I", !is.na(hla_typing)) %>% 
  separate_rows(hla_typing, sep = ",") %>% 
  dplyr::rename(hla_type = hla_typing) %>% distinct() %>% 
  group_by(dataset_id, ms_name) %>% 
  mutate(n_hla = n(), hla_len = nchar(hla_type), min_len = min(hla_len),
         hla_outlier = ifelse(str_detect(hla_type, "A2401|B4301|C1201"), T, F),
         has_hla_outlier = max(hla_outlier)) %>% 
  ungroup() %>% filter(min_len >= 5, has_hla_outlier == 0) %>% 
  select(-min_len, -hla_len, has_hla_outlier, hla_outlier) %>% 
  group_by(dataset_id, ms_name) %>% 
  mutate(hla_typing = paste0(hla_type, collapse = ",")) %>% 
  inner_join(peptide %>% 
               mutate(pep_len = nchar(peptide)) %>% 
               filter(pep_len >= 8, pep_len <= 12) %>% 
               select(peptide, dataset_id, ms_name, pep_len),
             c("ms_name", "dataset_id"), relationship = "many-to-many")

pep_can_nonc_hladf <- pep_can_nonc %>% semi_join(hla_df %>% distinct(peptide), 
                                                 "peptide")
setDT(hla_df)
hla_df <- hla_df[pep_can_nonc_hladf, on = 'peptide']
```

# Merge P-sites and save for Riboseq visualization (only once)

```{r}
bw_files <- list.files(path = file.path(raw_dir, "20240705_Gwisp_bw"), 
  pattern = paste0(".bedgraph"), full.names = TRUE)
bw_files_hht <- bw_files[which(str_detect(bw_files, "RiboProInit"))]
bw_files_ribo <- bw_files[which(!str_detect(bw_files, "RiboProInit"))]

hht_psites <- lapply(bw_files_hht, 
                     function(x) read_table(x, col_names = c("chr", "start", 
                                                             "end", "psites"))) %>% 
  bind_rows() %>% dplyr::count(chr, start, end, wt = psites)

ribo_psites <- lapply(bw_files_ribo, 
                     function(x) read_table(x, col_names = c("chr", "start", 
                                                             "end", "psites"))) %>% 
  bind_rows() %>% dplyr::count(chr, start, end, wt = psites)

saveRDS(ribo_psites, file.path(raw_dir, "ribo_psites.Rds"))
saveRDS(hht_psites, file.path(raw_dir, "hht_psites.Rds"))
```

# NetMHCpan predictions - All peptides

## Prepare input for HLA predictions

```{R}
netmhcpan_input <- pep_net_orf %>% distinct(peptide, hla_type) %>% 
  rename("allele" = 2) %>% mutate(
  allele = paste0("HLA-", str_extract(allele, "^[A-Z]\\d\\d"), ":", 
                  str_extract(allele, "(?<=^[A-Z]\\d\\d)\\d+"))
) %>% select(peptide, allele)

hla_df %>% distinct(hla_type) %>% arrange(hla_type) %>% unlist()

netmhcpan_input_s <- netmhcpan_input %>% split(netmhcpan_input$allele)
write_net_res <- sapply(1:length(netmhcpan_input_s), 
                        function(x) write_net(x, netmhcpan_input_s[[x]]))
write.table(names(netmhcpan_input_s), paste0(net_input, "/net_alleles.txt"), 
            row.names = F, col.names = F, quote = F)

# Run NetMHCpan
```

## Process prediction results
### Load netMHCpan data
#### Read from prediction files
```{r}
net_files <- list.files(path = net_output, 
  pattern = paste0("net_.*.xls"), full.names = TRUE)
netmhcpan <- lapply(net_files, read_netmhcpan) %>% bind_rows() 
```
#### Or use earlier generated data (from Github)
```{r}
netmhcpan <- readRDS(net_output_rds)
```

```{r}
setDT(hla_df)
setDT(netmhcpan)

peptide_net <- netmhcpan[hla_df, on = c("peptide", "hla_type"), nomatch = NULL]

pep_net_orf <- peptide_net %>% 
  group_by(peptide, dataset_id, ms_name, pep_len, sample_type, is_cancer, 
           is_cellline, pep_type) %>% 
  mutate(rank = as.numeric(rank)) %>% 
  summarise(min_rank = min(rank), hla_type = hla_type[which.min(rank)], 
            hla_typing = first(hla_typing),
            is_binding = ifelse(min_rank <= 2, T, F)) %>% ungroup() %>% 
  left_join(pep_nonc %>% distinct(peptide, orf_biotype), "peptide",
            relationship = "many-to-many") %>% 
  mutate(is_orf = ifelse(is.na(orf_biotype), F, T)) %>% ungroup()

n_dataset <- pep_net_orf %>% ungroup() %>% distinct(dataset_id) %>% nrow()
```

# Figure 2 - nonHLA figures

```{r}
nonhla_count <- table_s2 %>% select(-orf_biotype) %>% 
  inner_join(orf_sequences %>% select(orf_name, orf_biotype), "orf_name") %>% 
  dplyr::count(orf_name, orf_biotype) %>% group_by(orf_biotype) %>% 
  summarise(n_pep = sum(n), n_orf_1 = sum(n == 1), n_orf_2 = sum(n >= 2)) %>% 
  pivot_longer(c(n_orf_1, n_orf_2, n_pep)) %>% 
  mutate(x_label = str_replace_all(name, "_\\d", ""),
         name = factor(name, levels = c("n_pep", "n_orf_2", "n_orf_1"))) %>% 
  make_biotype_levels()

nonhla_count_group <- nonhla_count %>% filter(name != "n_pep") %>% 
  group_by(name) %>% summarise(value = sum(value))

val_count <- nonhla_count %>% dplyr::count(name, wt = value) %>% 
  inner_join(data.frame(name = c("n_orf_2", "n_orf_1"), n_val = c(30, 36)), "name") %>% 
  mutate(n_nonval = n - n_val) %>% select(-n) %>% column_to_rownames("name") %>%  
  t() %>% data.frame() %>% rownames_to_column("label") %>% 
  mutate(ypos2 = cumsum(n_orf_2)- 0.5*n_orf_2,
         ypos1 = cumsum(n_orf_1)- 0.5*n_orf_1)

nonhla_pep_count <- table_s3 %>% 
  filter(HPP.Category %in% c("HPP+", "1PepCandidate")) %>% 
  mutate(n_pep = ifelse(n_peptides_in_entry > 4, ">4", n_peptides_in_entry),
         n_pep = factor(n_pep, c(1,2,3,4, ">4"))) %>% 
  dplyr::count(n_pep)

figure_2A <- ggplot() +
  geom_bar(data = nonhla_count %>% filter(name == "n_pep"), 
           aes(x = orf_biotype, y = value, fill = name),
           stat = "identity", show.legend = F) + black_theme +
  scale_x_discrete(name = element_blank(), expand = c(0,0)) +
  scale_y_continuous(name = "Number of peptides", expand = c(0,0)) +
  scale_fill_manual(values = blue_col[3]) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_2B1 <- ggplot() +
  geom_bar(data = nonhla_count %>% filter(name != "n_pep"), 
           aes(x = orf_biotype, y = value, fill = name),
           stat = "identity", position = "stack") + black_theme +
  scale_x_discrete(name = element_blank(), expand = c(0,0)) +
  scale_y_continuous(name = "Number of ncORFs", expand = c(0,0)) +
  scale_fill_manual(values = green_col[c(6,1)],
                    labels = c("Multiple peptide detections",
                               "Single peptide detections"),
                    name = element_blank()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_2B2 <- ggplot() +
  geom_bar(data = nonhla_count_group, 
           aes(x = "", y = value, fill = name),
           stat = "identity", position = "stack", show.legend = F) + black_theme +
  scale_x_discrete(name = element_blank(), expand = c(0,0)) +
  scale_y_continuous(name = "Number of ncORFs", expand = c(0,0)) +
  scale_fill_manual(values = green_col[c(6,1)]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
        axis.ticks.x = element_blank())

figure_2C1 <- ggplot(val_count, aes(x = "", y = n_orf_2, fill = label)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar("y", start = 0, direction = 1) + 
  theme_void() +
  theme(legend.text = element_text(size = 7, colour = "black")) +
  geom_text(aes(y = ypos2, label = n_orf_2), color = "black", size = 7/.pt) +
  scale_fill_manual(values = c(pink_col[4], green_col[4]), 
                    name = element_blank(), 
                    labels = c("Insufficient\nevidence",
                               "Passing\ninspection"))

figure_2C2 <- ggplot(val_count, aes(x = "", y = n_orf_1, fill = label)) +
  geom_bar(stat = "identity", width = 1, color = "white") +
  coord_polar("y", start = 0, direction = 1) + 
  theme_void() +
  theme(legend.text = element_text(size = 7, colour = "black")) +
  geom_text(aes(y = ypos1, label = n_orf_1), color = "black", size = 7/.pt) +
  scale_fill_manual(values = c(pink_col[4], green_col[4]), 
                    name = element_blank(), 
                    labels = c("Insufficient\nevidence",
                               "Passing\ninspection"))

figure_2D <- ggplot(nonhla_pep_count, aes(y = n, x = n_pep)) + 
  geom_bar(position = "stack", stat = "identity", show.legend = F, 
           fill = green_col[4]) + black_theme +
  scale_x_discrete(name = "Number of distinct peptides", expand = c(0,0)) +
  scale_y_continuous(name = "Number of non-canonical ORFs", expand = c(0,0)) +
  geom_text(aes(label = n), size = 6/.pt)

figure_2 <- figure_2A + figure_2B1 + figure_2B2 + (figure_2C1 / figure_2C2) + 
  figure_2D + plot_layout(guides = 'collect', widths = c(7, 7, 1, 10, 5))

ggsave("Figure_2.svg", device = "svg", path = plot_dir, figure_2, 
       width = 2600, height = 900, units = "px")
```

# Figure 3 - HLA figures

## Figure 3A-D - Compare orf categories

```{r}
orf_pep_stats <- pep_nonc %>% rowwise() %>% 
  mutate(cov = paste0(start_loc:end_loc, collapse = ",")) %>% 
  group_by(orf_name) %>% 
  summarise(n_pep = n(), cov = paste0(cov, collapse = ",")) %>% rowwise() %>% 
  mutate(n_cov = cov %>% str_split(",", simplify = T) %>% as.numeric() %>% 
           na.omit() %>% unique() %>% length()) %>% 
  select(-cov) %>% 
  full_join(orf_sequences, "orf_name") %>% 
  replace_na(list(n_pep = 0, n_cov = 0)) %>% 
  mutate(orf_len = nchar(protein),
         perc_cov = n_cov / orf_len,
         cov_cat = cut(perc_cov,
                       breaks = seq(0,1, 0.1),
                       labels = sapply(seq(0,90,10), function(i) 
                         paste0(i, "-", i+10, "%")))) %>% 
  make_biotype_levels()
  
orf_pep_stats_count <- orf_pep_stats %>% group_by(orf_biotype) %>% 
  summarise(n_orf = sum(n_pep >= 1), n_pep = sum(n_pep)) %>% 
  pivot_longer(-orf_biotype) %>% 
  mutate(name = factor(name, levels = c("n_pep", "n_orf")))

orf_quality <- table_s5 %>% filter(n_peptides_in_entry >= 2) %>% 
  mutate(is_sufficient = Manual.inspection.of.GWIPS.Viz.RiboSeq.Data != 
           "Insufficient",
         study = ifelse(str_detect(identifier, "norep"), "1", ">1"),
         study = factor(study, levels = c("1", ">1"))) %>% 
  group_by(study) %>% dplyr::count(is_sufficient)

figure_3A <- ggplot(orf_pep_stats_count, 
                    aes(y = value, x = orf_biotype, fill = name)) + 
  geom_bar(position = position_dodge(), stat = "identity", show.legend = F) + 
  black_theme + 
  scale_x_discrete(name = element_blank(), expand = c(0,0)) +
  scale_y_continuous(name = "Number of detections", expand = c(0,0)) +
  scale_fill_manual(values = green_col[c(6,1)]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_3B <- ggplot(orf_pep_stats %>% dplyr::count(n_pep) %>% 
                      filter(n_pep > 0), aes(y = n, x = n_pep)) + 
  geom_bar(position = "stack", stat = "identity", show.legend = F, 
           fill = green_col[1]) + black_theme +
  scale_x_continuous(name = "Number of distinct peptides", expand = c(0,0),
                     breaks = 1:16) +
  scale_y_continuous(name = "Number of non-canonical ORFs", expand = c(0,0)) +
  geom_text(aes(label = n), size = 6/.pt)

figure_3C <- 
  ggplot(orf_pep_stats %>% filter(perc_cov > 0) %>% 
           mutate(n_pep = ifelse(n_pep > 1, "2", "1")), 
         aes(perc_cov, orf_len, color = n_pep)) +
  geom_point(show.legend = F) + black_theme +
  geom_smooth(method = "loess", formula = y ~ x, se = T, 
              mapping = aes(color = n_pep), linewidth = 1,
              show.legend = F) +
  scale_x_continuous(name = "Percentage aa covered", expand = c(0,0),
                     limits = c(0,1), breaks = c(0,.25,.5,.75,1),
                     labels = c("0%", "25%", "50%", "75%", "100%")) +
  scale_y_continuous(name = "ncORF length (aa)", limits = c(0, NA),
                     expand = c(0,0)) +
  scale_color_manual(name = element_blank(), values = c(green_col[1], blue_col[4]),
                     labels = c("Detected by single peptide", "Detected by multiple peptides"))

figure_3D <- ggplot(orf_quality, aes(y = n, x = study, fill = is_sufficient)) + 
  geom_bar(position  = "stack", stat = "identity", show.legend = F) + 
  black_theme + 
  scale_x_discrete(name = "Number of studies\nin which ncORF\nwas detected", 
                   expand = c(0,0)) +
  scale_y_continuous(name = "Number of non-canonical ORFs", expand = c(0,0)) +
  scale_fill_manual(values = c(grey_col[7], green_col[1])) 

figure_3AD <- figure_3A + figure_3B + figure_3C + figure_3D + 
  plot_layout(nrow = 1, widths = c(1.6,2.6,1.4,1))

ggsave("Figure_3AD.svg", device = "svg", path = plot_dir, figure_3AD, 
       width = 2500, height = 700, units = "px")

# Statistical test for figure 3C
wilcox.test(data = orf_pep_stats %>% filter(perc_cov > 0) %>% 
              mutate(n_pep = ifelse(n_pep > 1, "2", "1")), 
            orf_len ~ n_pep)$p.value

orf_pep_stats %>% filter(perc_cov > 0) %>% mutate(n_pep = ifelse(n_pep > 1, "2", "1")) %>% 
  group_by(n_pep) %>% summarize(mean_len = mean(orf_len))
```

## Figure 3E-F - Binding predictions

```{r}
avg_pep_len <- peptide_hla %>% filter(hla_class == "I") %>% ungroup() %>% 
  mutate(p_len = nchar(peptide)) %>% 
  group_by(dataset_id, ms_name) %>% 
  summarise(avg_len = mean(p_len), pep_count = n())

peplen_plot_data <- pep_net_orf %>% 
  group_by(dataset_id, ms_name, is_binding) %>% 
  summarise(n = n(), orf_count = sum(is_orf)) %>% 
  mutate(is_binding = ifelse(is_binding, "binding", "non_binding")) %>% 
  pivot_wider(names_from = is_binding, values_from = c(n, orf_count), 
              values_fill = 0) %>% 
  inner_join(avg_pep_len, c("dataset_id", "ms_name")) %>% 
  mutate(n = n_non_binding + n_binding,
         perc = n_binding / n,
         orf_count = orf_count_non_binding + orf_count_binding,
         n_non_can_perc = orf_count / n*100,
         n_non_can_perc_lim = ifelse(n_non_can_perc > 3, 3, n_non_can_perc)) %>% 
  arrange(n_non_can_perc) %>% filter(avg_len < 20)

figure_3E1 <- 
  ggplot(peplen_plot_data,aes(x = avg_len, y = perc, color = n_non_can_perc_lim)) + 
  geom_point(aes(size = pep_count), stroke = 0) + black_theme +
  scale_color_gradient(name = "% non-can\npeptides",
                      low = grey_col[1], high = red_col,
                      limits = c(0, 3),
                      breaks = c(0:3),
                      labels = c("0%", "1%", "2%", "3 - 25%")) +
  scale_size_continuous(breaks = c(4000, 8000, 12000), 
                        labels = c("4,000", "8,000", "12,000"),
                        name = "# Total peptides") +
  scale_y_continuous(limits = c(0,1.01), 
                     name = "% binders (rank ≤ 2)\namongst peptides from 8-12 aa", 
                     breaks = c(0, 0.25, 0.5, 0.75, 1),
                     labels = c("0%", "25%", "50%", "75%", "100%")) +
  scale_x_reverse(limits = c(19.3,8.7), 
                     name = "Mean peptide length (aa)", 
                     breaks = c(10, 12, 14, 16, 18)) +
  geom_vline(xintercept = 12, linetype = "dashed")

figure_3E2 <- ggplot(peplen_plot_data %>% filter(orf_count > 0), 
                          aes(x = avg_len, y = perc, color = n_non_can_perc_lim)) + 
  geom_point(aes(size = orf_count), stroke = 0) + 
  black_theme + 
  theme(legend.direction = "vertical", legend.box = "horizontal") +
  scale_color_gradient(name = "% non-can\npeptides",
                      low = grey_col[1], high = red_col,
                      limits = c(0, 3),
                      breaks = c(0:3),
                      labels = c("0%", "1%", "2%", "3 - 25%")) +
  scale_size_continuous(breaks = c(20, 40, 60),
                        name = "# non-can\npeptides") +
  scale_y_continuous(limits = c(0,1.01), 
                     name = "% binders (rank ≤ 2)\namongst peptides from 8-12 aa", 
                     breaks = c(0, 0.25, 0.5, 0.75, 1),
                     labels = c("0%", "25%", "50%", "75%", "100%")) +
  scale_x_continuous(limits = c(8.7,19.3), 
                     name = "Mean peptide length (aa)",
                     breaks = c(10, 12, 14, 16, 18)) +
  geom_vline(xintercept = 12, linetype = "dashed")



bind_cmp_plot_data <- pep_net_orf %>% 
  dplyr::count(dataset_id, pep_type, is_binding) %>% 
  mutate(is_binding = ifelse(is_binding, "b", "nb")) %>% 
  pivot_wider(names_from = c(pep_type, is_binding), values_from = n, 
              values_fill = 0) %>% rowwise() %>% 
  mutate(n = sum(canonical_b, canonical_nb, noncanonical_b, noncanonical_nb, 
                 other_b, other_nb),
         total_non_can = noncanonical_b + noncanonical_nb,
         perc_can = canonical_b / (canonical_b + canonical_nb),
         perc_non_can = noncanonical_b / total_non_can,
         n_non_can_perc = total_non_can / n*100) %>% 
  arrange(n_non_can_perc)
  
figure_3F <- ggplot(bind_cmp_plot_data %>% filter(!is.nan(perc_non_can)),
                        aes(x = perc_can, y = perc_non_can, color = n_non_can_perc)) + 
  geom_point(aes(size = total_non_can), stroke = 0) + 
  black_theme +
  scale_color_gradient(name = "% non-can\npeptides",
                      low = grey_col[1], high = red_col,
                      limits = c(0, 3), breaks = c(0:3),
                      labels = c("0%", "1%", "2%", "3%")) +
  scale_y_continuous(limits = c(0,1),
                     name = "% binders non-canonical\namongst peptides from 8-12 aa") +
  scale_x_continuous(limits = c(0,1),
                     name = "% binders canonical\namongst peptides from 8-12 aa")

figure_3EF <- figure_3E1 + figure_3E2 + figure_3F + plot_layout(nrow = 1)
  
ggsave("Figure_3EF.svg", device = "svg", path = plot_dir, figure_3EF, 
       width = 3152.30, height = 682.48, units = "px")
```

## Figure 3G - Heatmap

```{r}
# Heatmap colors
ht_colors <- colorRamp2(breaks = c(-1, 0, 2,2.000001, 100),
                        colors = c("white", "#0072B2", "#0072B2", 
                                   "#E69F00", "#E69F00"))

# Heatmap data
ht_data <- pep_net_orf %>% ungroup() %>% filter(is_orf) %>% 
  distinct(peptide, min_rank, hla_typing, orf_biotype, sample_type) %>% 
  mutate(pep = paste0(peptide, "_", orf_biotype),
         hla_sample = paste0(hla_typing, "_", sample_type))

ht_all <- ht_data %>% distinct(pep, hla_typing, min_rank, orf_biotype) %>% 
  pivot_wider(names_from = hla_typing, values_from = min_rank, 
              values_fill = -1) %>% column_to_rownames("pep")

ht_tis <- ht_data %>% select(pep, hla_sample, min_rank, orf_biotype) %>% 
  pivot_wider(names_from = hla_sample, values_from = min_rank, 
              values_fill = -1) %>% column_to_rownames("pep")

ht_mtr_all <- ht_all %>% select(-orf_biotype) %>% as.matrix()
ht_mtr_tis <- ht_tis %>% select(-orf_biotype) %>% as.matrix()

# Concordance predictions grouped by HLA typing
concordance <- ht_tis %>% select(-orf_biotype) %>% 
  pivot_longer(c(1:ncol(.))) %>% filter(value > -1) %>% 
  mutate(binding = value <= 2) %>% dplyr::count(binding) %>% 
  pivot_wider(names_from = binding, values_from = n) %>% 
  mutate(perc = `TRUE` / (`TRUE` + `FALSE`) * 100) %>% pull(perc)

# Row and column splits
ht_row_split <- ht_tis %>% select(orf_biotype) %>% make_biotype_levels()
  
ht_col_split <- data.frame(hla_sample = 
                             colnames(ht_tis %>% select(-orf_biotype))) %>% 
  mutate(col_split = factor(str_extract(hla_sample, "(?<=\\d_).*"), 
                            levels = c("Non-cancer_cell line", 
                                       "Non-cancer_non-cell line", 
                                       "Cancer_cell line", 
                                       "Cancer_non-cell line"))) %>% 
  column_to_rownames("hla_sample")

# Column annotation
col_counts_tis <- apply(ht_mtr_tis, 2, function(x) sum(x <= 2 & x >= 0))

pep_per_hla_tis <- pep_net_orf %>% ungroup %>% 
  distinct(peptide, hla_typing, sample_type) %>% 
  mutate(hla_sample = paste0(hla_typing, "_", sample_type)) %>% 
  dplyr::count(hla_sample) %>% 
  semi_join(data.frame(hla_sample = colnames(ht_mtr_tis)), "hla_sample") %>% 
  arrange(match(hla_sample, colnames(ht_mtr_tis))) %>% 
  column_to_rownames("hla_sample")
  
color_ann_df_tis <- ht_col_split %>% 
  mutate(is_cancer = str_extract(col_split, ".*(?=_)"),
         is_cellline = str_extract(col_split, "(?<=_).*")) %>% 
  select(-col_split)

col_ann_colours <- list(is_cancer = c('Cancer' = blue_col[2], 
                                      'Non-cancer' = blue_col[6]),
  is_cellline = c('cell line' = green_col[2],
                  'non-cell line' = green_col[6]))

# Column annotation
col_ha_tis <- HeatmapAnnotation(
  df = color_ann_df_tis,
  binders = anno_barplot(col_counts_tis, border = F, height = unit(6, "mm"),
                         axis_param = list(gp = gpar(fontsize = 7), 
                                           at = c(0, 200))),
  total_pep = anno_barplot(pep_per_hla_tis, border = F, height = unit(6, "mm"),
                           axis_param = list(at = c(0, 2e4),
                                             labels = c("0", "20k"),
                                             gp = gpar(fontsize = 7))),
  col = col_ann_colours, show_annotation_name = F,
  annotation_height = unit(c(2,2,6,6), "mm"),
  annotation_legend_param = 
    list(is_cancer = list(title = "",
                          labels = c("Cancer", "Non-cancer"),
                          at = c("Cancer", "Non-cancer"),
                          labels_gp = gpar(fontsize = 7)),
         is_cellline = list(title = "", 
                          labels = c("Cell line", "Non-cell line"),
                          at = c("cell line", "non-cell line"),
                          labels_gp = gpar(fontsize = 7))),
  gap = unit(c(.3, .3, 2, 1.5), "mm"))

# Row annotation
row_counts <- apply(ht_mtr_all, 1, function(x) sum(x <= 2 & x >= 0))
row_counts_nb <- apply(ht_mtr_all, 1, function(x) sum(x > 2))

row_ha <- HeatmapAnnotation(
  binders = anno_barplot(row_counts, border = F, gp = gpar(col = "#0072B2"),
                         ylim = c(0,120), extend = 0, width = unit(8.4, "mm"),
                         axis_param = list(gp = gpar(fontsize = 7), 
                                           at = c(0, 50 ,100))),
  non_binders = anno_barplot(row_counts_nb, border = F, 
                             gp = gpar(col = "#E69F00"), ylim = c(0,50), 
                             extend = 0, width = unit(3.5, "mm"), 
                             axis_param = list(gp = gpar(fontsize = 7), 
                                               at = c(0,50))),
  show_annotation_name = F, which = "row", gap = unit(c(1.5, 1.5), "mm"))

# Heatmap
figure_3G <- Heatmap(ht_mtr_tis, show_row_names = F, 
                       show_column_names = F, 
        cluster_columns = T, cluster_rows = T, col = ht_colors,
        column_split = ht_col_split, use_raster = F,
        row_split = ht_row_split, row_title_rot = 0,
        top_annotation = col_ha_tis, right_annotation = row_ha,
        cluster_row_slices = F, 
        cluster_column_slices = F, 
        show_row_dend = F, show_column_dend = F,  show_heatmap_legend = F,
        column_title_side = "bottom",
        column_title = paste0("Samples grouped by ", ncol(ht_mtr_all), 
                              " HLA typings and 4 sample types (n = ",
                              ncol(ht_mtr_tis), " combinations)"),
        column_title_gp = gpar(fontsize = 7),
        row_title_gp = gpar(fontsize = 7),
        height = nrow(ht_mtr_tis)*unit(0.030, "mm"),
        width = ncol(ht_mtr_tis)*unit(0.3, "mm"))

svg(file = file.path(plot_dir, "Figure_3G.svg"))
draw(figure_3G, annotation_legend_side = "bottom", newpage = F)
dev.off()
```

# Figure 4 - Detection determinants

## Figure 4A - Undetected vs detected

```{r}
seq_stats_properties <- seq_stats %>% dplyr::rename(full_len = len) %>% 
  mutate(to_duplicate = ifelse(pro_cat != "Canonical", 3, 1)) %>% 
  uncount(to_duplicate) %>% group_by(protein_accession) %>% 
  mutate(
    pro_cat = case_when(
      row_number() == 1 ~ paste0(pro_cat, "__full"),
      row_number() == 2 ~ paste0(pro_cat, "__start"),
      row_number() == 3 ~ paste0(pro_cat, "__end")),
    protein = case_when(
      row_number() == 1 ~ protein,
      # Add 0.001 to round to ensure .5 is always rounded up
      row_number() == 2 ~ str_sub(protein, 2, round(full_len*0.3+0.001) + 1),
      row_number() == 3 ~ str_sub(protein, 
                                  full_len - round(full_len*0.3+0.001) + 1, 
                                  full_len))) %>% 
  get_stats() %>% group_by(protein_accession) %>% 
  filter(len >= 16) %>% select(-protein, -full_len) %>% ungroup() %>% 
  mutate(order_ids = ifelse(detected == "Detected", "1", "2"),
         fill_group = paste0(pro_cat, "__", detected))

l_lab <- "Sequence\nlength (aa)"
h_lab <- "Hydrophobicity\n(Kyte-Doolittle)"
p_lab <- "Isoelectric\npoint"
v_cats <- c("Non-canonical__full", "Canonical__full")

#TODO remove y_val from perform w_test
figure_4A1 <- lapply(1:2, function(i) {
  cat <- v_cats[i]
  limits <- list(c(NA, NA), c(NA, NA))
  plot_data <- seq_stats_properties %>% filter(pro_cat == cat) %>% 
    mutate(y_val = len)
  p_val <- perform_wtest(plot_data, 8)
  create_violin(plot_data, l_lab, p_val, limits[[i]], 
                n_colors = c((4*i-1):(4*i)), log = T)
})
figure_4A2 <- lapply(1:2, function(i) {
  cat <- v_cats[i]
  plot_data <- seq_stats_properties %>% filter(pro_cat == cat) %>% 
    mutate(y_val = hydro)
  p_val <- perform_wtest(plot_data, 8)
  create_violin(plot_data, h_lab, p_val, c(-4.1, 2.5), 
                n_colors = c((4*i-1):(4*i)))
})
figure_4A3 <- lapply(1:2, function(i) {
  cat <- v_cats[i]
  plot_data <- seq_stats_properties %>% filter(pro_cat == cat) %>% 
    mutate(y_val = pi)
  p_val <- perform_wtest(plot_data, 8)
  create_violin(plot_data, p_lab, p_val, c(2.9, 14), 
                n_colors = c((4*i-1):(4*i)))
})

figure_4A_plots <- lapply(1:6, function(i) {
  one_plot <- c(figure_4A1, figure_4A2, figure_4A3)[[i]]
  if (i <= 4) {
    one_plot <- one_plot + remove_x
  }
  if (i %in% c(4,6)) {
    one_plot <- one_plot + remove_y
  }
  if ( i == 2) {
    one_plot <- one_plot + theme(axis.title.y = element_blank())
  }
  return(one_plot)
})
  
  
figure_4A <- wrap_plots(figure_4A_plots, ncol = 2)

ggsave("Figure_4A.svg", device = "svg", path = plot_dir, figure_4A, 
       width = 1250, height = 1750, units = "px")
```

## Figure 4B - Hydrophobicity end of sequence

```{r}
seq_stats_hydro <- seq_stats %>% 
  mutate(protein = stri_reverse(protein),
         duplicate = 101) %>%
  uncount(duplicate) %>% group_by(protein_accession) %>% 
  mutate(start = row_number(), end = start + 14,
         protein = str_sub(protein, start, end),
         pro_len = nchar(protein),
         c_offset = -(start - 1),
         hydro = hydrophobicity(protein, "KyteDoolittle")) %>% 
  filter(pro_len >= 15) 

seq_stats_hydro_grouped <- seq_stats_hydro %>% 
  group_by(orf_biotype, c_offset) %>% 
  mutate(low_conf = t.test(hydro)$conf.int[1], 
         high_conf = t.test(hydro)$conf.int[2]) %>% 
  summarise(hydro = mean(hydro), low_conf = first(low_conf),
            high_conf = first(high_conf)) %>% 
  filter(!orf_biotype %in% c("doORFs", "Processed transcript ORFs"))

seq_stats_hydro_grouped_order <- seq_stats_hydro_grouped %>% 
  filter(c_offset == 0) %>% arrange(-hydro) %>% pull(orf_biotype)

seq_stats_hydro_grouped <- seq_stats_hydro_grouped %>% 
  mutate(orf_biotype = factor(orf_biotype, 
                              levels = seq_stats_hydro_grouped_order))

figure_4B <- ggplot(seq_stats_hydro_grouped, 
       aes(x = c_offset, y = hydro, group = orf_biotype)) + 
  geom_linerange(aes(ymin = low_conf, ymax=  high_conf, color = orf_biotype), 
                 alpha = 0.5, linewidth = 0.3) +
  geom_point(mapping = aes(color = orf_biotype), size = .2) + 
  geom_smooth(method = "loess", formula = y ~ x, se = F, 
              mapping = aes(color = orf_biotype), linewidth = .7)  +
  scale_x_continuous(expand = c(0,0), 
                     name = "Position relative to C-terminus") +
  scale_y_continuous(limits = c(-1,0.5), expand = c(0,0),
                     name = "Hydrophobicity (Kyte-Doolittle)") +
  scale_color_manual(values = brewer.pal(6, "Dark2"),
                     name = "ORF/protein category") +
  black_theme

ggsave("Figure_4B.svg", device = "svg", path = plot_dir, figure_4B, 
       width = 1250, height = 1250, units = "px")
```

## Figure 4C - Expression

```{r}
gtex_df <- gtex_data %>% 
  dplyr::rename(gene_id = X) %>% pivot_longer(-gene_id) %>% 
  group_by(gene_id) %>% summarise(mean_fpkm = mean(value))

orf_gtex <- seq_stats %>% 
  inner_join(orf_sequences %>% 
               select(protein_accession, gene_id), "protein_accession") %>% 
  inner_join(gtex_df, "gene_id") %>% mutate(y_val = mean_fpkm)%>% 
  mutate(order_ids = ifelse(detected == "Detected", "1", "2"),
         fill_group = paste0(pro_cat, "__", detected)) 

# Number of missing ncORFs genes in GTEx
orf_sequences %>% anti_join(orf_gtex, "protein_accession") %>% nrow()

# Mean FPKM statistics
orf_gtex %>% group_by(detected) %>% summarise(mean_y = mean(y_val))

p_val_gtex <- perform_wtest(orf_gtex, 1)
figure_4C <- create_violin(orf_gtex %>% filter(y_val > 0), "Mean FPKM", 
                                   p_val_gtex, c(0, 1600), n_colors = c(3,4), 
                                   pseudo_log = T, 
                                   y_breaks = c(0,10,250,500,1000,2000))
  

ggsave("Figure_4C.svg", device = "svg", path = plot_dir, figure_4C, 
       width = 625, height = 600, units = "px")
```

## Figure 4D - Peptide location

```{r}
pep_pos_can <- pep_can %>% inner_join(can_id, "protein_accession") %>% 
  inner_join(can_nonc_seq, "protein_accession") %>% 
  semi_join(peptide_hla %>% filter(hla_class == "I"), "peptide") %>% 
  distinct(protein, peptide, start_loc, end_loc) %>% 
  mutate(len = nchar(protein),
         start_offset = start_loc - 1,
         end_offset = -(len - end_loc))

pep_pos_nonc <- pep_nonc %>% 
  semi_join(peptide_hla %>% filter(hla_class == "I"), "peptide") %>% 
  distinct(protein, peptide, start_loc, end_loc) %>% 
  mutate(len = nchar(protein),
         start_offset = start_loc - 1,
         end_offset = -(len - end_loc))

can_lim = c(0, pep_pos_can %>% filter(end_offset == 0) %>% nrow())
nonc_lim = c(0, pep_pos_nonc %>% filter(end_offset == 0) %>% nrow())

# Compute statistics on fold changes
start_fc <- lapply(list(pep_pos_can, pep_pos_nonc), function(i) 
  process_data(i, start_offset <= 1, "start_offset", "1", "0")) %>% 
  bind_rows() %>% perform_fisher_test()

end_fc <- lapply(list(pep_pos_can, pep_pos_nonc), function(i) 
  process_data(i, end_offset >= -1, "end_offset", "0", "-1")) %>% 
  bind_rows() %>% perform_fisher_test()

can_start <- create_histogram(
  pep_pos_can %>% filter(start_offset <= 25),
  start_offset, pink_col[1],
  y_label = "Number of\ncanonical peptides",
  y_limits = can_lim, y_breaks = c(0, 5000, 10000),
  y_labels = c(0, "5k", "10k")
)

can_end <- create_histogram(
  pep_pos_can %>% filter(end_offset >= -25),
  end_offset, pink_col[1],
  y_limits = can_lim
)

nonc_start <- create_histogram(
  pep_pos_nonc %>% filter(start_offset <= 25),
  start_offset, green_col[1],
  y_label = "Number of\nncORF peptides",
  x_label = "Position relative to N-terminus (aa)",
  y_limits = nonc_lim, y_breaks = c(0, 250, 500)
)

nonc_end <- create_histogram(
  pep_pos_nonc %>% filter(end_offset >= -25),
  end_offset, green_col[1],
  x_label = "Position relative to C-terminus (aa)",
  y_limits = nonc_lim
)

figure_4D <- (can_start + can_end) / (nonc_start + nonc_end)

ggsave("Figure_4D.svg", device = "svg", path = plot_dir, figure_4D,
       width = 1100, height = 800, units = "px")
```

## Figure 4E - Ligand Atlas

```{r}
ligand_atlas_data <-  pep_net_orf %>% ungroup() %>% 
  filter(dataset_id == "PXD019643", is_binding, pep_type != "other") %>% 
  mutate(tissue_type = str_extract(ms_name, "(?<=_)\\w+(?=_W6)"),
         tissue_type = str_replace(tissue_type, "(?<=\\w)(?=[A-Z])", " ")) %>% 
  distinct(tissue_type, is_orf, peptide) %>% 
  dplyr::count(tissue_type, is_orf) %>% 
  mutate(is_orf = ifelse(is_orf, "ncORF", "CDS"),
         tissue_type = case_when(
           tissue_type == "Brain" ~ "Brain (Cerebral cortex)",
           tissue_type == "Mamma" ~ "Mamma (Breast)",
           tissue_type == "Myelon" ~ "Myelon (Spinal cord)",
           T ~ tissue_type)) %>% 
  pivot_wider(names_from = is_orf, values_from = n, values_fill = 0) %>% 
  mutate(total_ncORF = sum(.$ncORF), total_CDS = sum(.$CDS),
         ncORF_nottis = total_ncORF - ncORF,
         CDS_nottis = total_CDS - CDS) %>% 
  rowwise() %>% 
  mutate(p_val = 
           fisher_test_ligand_atlas(ncORF, ncORF_nottis, CDS, CDS_nottis)) %>% 
  ungroup() %>% mutate(p_adj = p_val * nrow(.),
                       is_sig = p_adj < 0.05) %>% rowwise() %>% 
  mutate(perc_orf = ncORF/(ncORF + CDS) * 100,
         conf_low = binom.test(ncORF,ncORF + CDS)$conf.int[1]*100,
         conf_high = binom.test(ncORF,ncORF + CDS)$conf.int[2]*100) %>% 
  ungroup() %>% arrange(perc_orf) %>% 
  mutate(tissue_type = factor(tissue_type, levels = tissue_type))

figure_4D1 <- ggplot() + 
  geom_bar(data = ligand_atlas_data, mapping = aes(tissue_type, ncORF),
           stat = "identity", fill = green_col[1]) +
  black_theme +
  scale_x_discrete(name = element_blank(), labels = c(), expand = c(0,0)) + 
  scale_y_continuous(name = "Number of\nncORF peptides", expand = c(0,0)) +
  theme(axis.ticks.x = element_blank())

figure_4D2 <- ggplot() + 
  geom_bar(data = ligand_atlas_data, mapping = aes(tissue_type, CDS),
           stat = "identity", fill = pink_col[1]) +
  black_theme +
  scale_x_discrete(name = element_blank(), labels = c(), expand = c(0,0)) + 
  scale_y_continuous(name = "Number of\nCDS peptides", expand = c(0,0),
                     breaks = c(0, 5000, 10000), labels = c("0", "5k", "10k")) +
  theme(axis.ticks.x = element_blank())

figure_4D3 <- ggplot(ligand_atlas_data, 
            mapping = aes(tissue_type, perc_orf, fill = is_sig)) +
  geom_bar(stat = "identity") + black_theme +
  scale_x_discrete(expand = c(0,0), name = element_blank()) + 
  scale_y_continuous(expand = c(0,0),
                     name = "% ncORF peptides") +
  scale_fill_manual(values = c(green_col[4], red_col), name = element_blank(),
                    labels = c("", "Significant\nenrichment")) +
  theme(axis.text.x = element_text(angle = 60,  hjust = 1)) +
  geom_hline(yintercept = sum(ligand_atlas_data$ncORF)/
               (sum(ligand_atlas_data$ncORF)+sum(ligand_atlas_data$CDS))*100, 
             linetype = "dashed")

figure_4E <- figure_4E1 / figure_4E2 / figure_4E3 

ggsave("Figure_4E.svg", device = "svg", path = plot_dir, figure_4E,
        width = 1544, height = 1375, units = "px")
```

# Figure 5 - Tiers

## Figure 5B - Tier assignment

```{r}
tiers <- table_s3 %>% select(orf_name, initial_tier, final_tier) %>% 
  bind_rows(table_s5 %>% select(orf_name, initial_tier, final_tier)) %>% 
  mutate(initial_tier = str_replace(initial_tier, " - .+", ""),
         final_tier = str_replace(final_tier, " - .+", ""),
         final_tier = str_replace(final_tier, ",.+", "")) %>% 
  distinct(orf_name, initial_tier, final_tier) %>% 
  right_join(orf_sequences %>% select(orf_name), "orf_name") %>% 
  replace_na(list(initial_tier = "Tier 4", final_tier = "Tier 4")) %>% 
  mutate(
  initial_tier = 
    factor(initial_tier, levels = c("Tier 1A", "Tier 1B", "Tier 2A", 
                                    "Tier 2B", "Tier 3", "Tier 4")),
  final_tier = 
    factor(final_tier, levels = c("Tier 4", "Tier 3", "Tier 2B", "Tier 2A", 
                                  "Tier 1B", "Tier 1A")))

tier_counts <- tiers %>% dplyr::count(initial_tier, final_tier) %>% 
  mutate(n = format(n , big.mark = ","))

figure_5B <- ggplot() +
  geom_tile(data = tier_counts, aes(initial_tier, final_tier), 
            color = "black", fill = "white") +
  geom_text(data = tier_counts, aes(initial_tier, final_tier, label = n), 
            size = 7/.pt) +
  scale_x_discrete(position = "top", name = "Provisional tier") +
  scale_y_discrete(name = "Final tier") +
  geom_text(data = tiers %>% dplyr::count(initial_tier) %>% 
              mutate(n = format(n , big.mark = ",")), 
            aes(x = initial_tier, y = 0.3, label = n), size = 7/.pt, 
            fontface='bold') +
  geom_text(data = tiers %>% dplyr::count(final_tier) %>% 
              mutate(n = format(n , big.mark = ",")), 
            aes(x = 5.6, y = final_tier, label = n), size = 7/.pt, hjust = 0, 
            fontface='bold') +
  black_theme +
  theme(axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_text(face='bold'),
        axis.text = element_text(face='bold')) +
  coord_cartesian(clip = "off", expand = FALSE)

ggsave("Figure_5B.svg", device = "svg", path = plot_dir, figure_5B, 
       width = 1000, height = 500, units = "px")
```

## Figure 5C - Tier 1A breakdown

```{r}
data_5c <- tiers %>% filter(initial_tier == "Tier 1A") %>% 
  mutate(final_tier = ifelse(is.na(final_tier) | final_tier == "Tier 3",
                                   "Tier 1A", as.character(final_tier)),
         final_tier = 
           factor(final_tier, levels = c("Tier 1B", "Tier 2A", 
                                         "Tier 2B", "Tier 4", "Tier 1A"))) %>% 
  dplyr::count(final_tier)
  
figure_5c <- ggplot(data_5c, aes(x = "", y = n, fill = final_tier)) +
  geom_bar(stat = "identity", width = 1, color = "white", show.legend = F) +
  coord_polar("y", start = 0, direction = 1) + 
  theme_void() +
  scale_fill_manual(values = c("#B9B1A4", "#C8C1B7","#D8D3CC" ,pink_col[4], 
                               green_col[4]))

ggsave("Figure_5c.svg", device = "svg", path = plot_dir, figure_5c, 
       width = 900, height = 900, units = "px")
```

# Figure 6 - ORF examples

## Figure 6A - nonHLA c11riboseqorf4

```{r}
b_w <- 0.7
orf_id <- "CONTRIB_GENCODE_c11riboseqorf4"

orf_seq <- can_nonc_seq %>% 
  filter(str_detect(protein_accession, "^CONTRIB_GENCODE_c11riboseqorf4$")) %>% 
  pull(protein)

nonhla_pep <- table_s2 %>% 
  filter(str_detect(identifier, "CONTRIB_GENCODE_c11riboseqorf4")) %>% 
  select(sequence, length) %>% rowwise() %>% 
  mutate(pep_start = gregexpr(sequence, orf_seq)[[1]],
         pep_end = pep_start + length - 1) %>% 
  ungroup() %>% 
  arrange(pep_start, pep_end) %>% 
  mutate(x_group = c(1, 2, 2, 3, 4, 4, 4, 5, 5, 5, 6)) %>% 
  group_by(x_group) %>%
  mutate(y_order = row_number(),
         y_min = y_order - 0.5 * b_w,
         y_max = y_order + 0.5 * b_w)

figure_6A <- ggplot() +
  geom_rect(nonhla_pep, 
            mapping = aes(xmin = pep_start, xmax = pep_end,
                          ymin = y_min, ymax = y_max), fill = "#42AB5C") +
  black_theme +
  scale_x_continuous(limits = c(1, nchar(orf_seq)), expand = c(0,0), 
                     breaks = c(1:nchar(orf_seq)),
                     labels = sapply(1:nchar(orf_seq), 
                                     function(x) ifelse(x %% 10 == 0, x, "")),
                     name = orf_seq) +
  scale_y_continuous(breaks = c()) +
  theme(axis.line.y = element_blank())

ggsave("Figure_6A.svg", device = "svg", path = plot_dir, figure_6A, 
       width = 2200, height = 300, units = "px")
```

## Figure 6B - HLA c17norep146

### Create c17norep146 kmers

```{r}
seq_17_146 <- can_nonc_seq %>% 
  filter(protein_accession == "CONTRIB_GENCODE_c17norep146") %>% 
  pull(protein)

seq_17_146_kmer_df<- sapply(8:12, function(i) extract_kmers(seq_17_146, i)) %>% 
  unlist() %>% enframe() %>% dplyr::rename(kmer = value) %>% select(-name)
```

### Run NetMHCpan (only once)

```{r}
unique_kmers <- seq_17_146_kmer_df %>% distinct(kmer) %>% rename("peptide" = 1)
write_net("", unique_kmers)

found_alleles <- hla_df %>% select(hla_type) %>% distinct() %>% 
  arrange(hla_type) %>% mutate(
  hla_type = paste0("HLA-", str_extract(hla_type, "^[A-Z]\\d\\d"), ":", 
                  str_extract(hla_type, "(?<=^[A-Z]\\d\\d)\\d+"))
)

write.table(found_alleles$hla_type, paste0(net_input, "/net_alleles.txt"), 
            row.names = F, col.names = F, quote = F)

# Run NetMHCpan
```

### Read NetMHCpan results
#### Either read prediction results
```{r}
# Read and process netmhcpan files
net_c17_146 <- list.files(path = file.path(processed_dir, 
                                          "netmhcpan_kmers_c7norep146"), 
                         pattern = "net_.*.xls", full.names = TRUE) %>%
  map_df(read_netmhcpan)
```
#### Or use pre-generated data (from github)
```{r}
net_c17_146 <- readRDS(file.path(raw_dir, net_c17_146.RDS)
```

### Create Figure 6B
```{r}
# Process alleles
alleles_17_146 <- pep_nonc %>% 
  filter(protein_accession == "CONTRIB_GENCODE_c17norep146") %>% 
  inner_join(hla_df, "peptide") %>% 
  separate_rows(hla_typing, sep = ",") %>% 
  dplyr::count(hla_typing)

# Process kmer data
seq_17_146_kmer_df_rank <- seq_17_146_kmer_df %>% 
  mutate(pep_len = nchar(kmer)) %>% group_by(pep_len) %>% 
  mutate(kmer_start = row_number(),
         kmer_end = row_number() + pep_len - 1) %>% ungroup() %>% 
  left_join(net_c17_146 %>% 
              inner_join(alleles_17_146, c("hla_type" = "hla_typing")), 
            c("kmer" = "peptide")) %>% 
  filter(rank <= 2) %>% 
  group_by(kmer, pep_len, kmer_start, kmer_end) %>% 
  summarize(min_rank = min(rank), .groups = "drop")

# Process contour data
seq_17_146_contour <- seq_17_146_kmer_df_rank %>%
  rowwise() %>% 
  mutate(amino_acids = list(kmer_start:kmer_end)) %>%
  unnest(amino_acids) %>%
  dplyr::count(amino_acids) %>% 
  right_join(tibble(amino_acids = 1:nchar(seq_17_146)), by = "amino_acids") %>% 
  replace_na(list(n = 0)) %>% 
  mutate(amino_acids = amino_acids + 0.5) %>% 
  add_row(amino_acids = 1, n = 0, .before = 1) %>% 
  mutate(n = ifelse(amino_acids == 1, lead(n), n),
         amino_acids = ifelse(amino_acids == 100.5, 100, amino_acids))

# Process kmer plot data
kmer_17_146_p_data <- seq_17_146_kmer_df_rank %>% 
  group_by(kmer_end) %>%
  summarise(min_start = min(kmer_start),
            kmer_start = list(kmer_start),
            pep_num = n()) %>%
  mutate(x_group = rep(1:3, c(8, 6, 5))) %>%
  group_by(x_group) %>%
  mutate(y_order = row_number()) %>% 
  unnest(kmer_start) %>% 
  
  mutate(pep_len = factor(kmer_end - kmer_start + 1, levels = 12:8)) %>% 
  group_by(kmer_end) %>% 
  arrange(pep_len) %>% 
  mutate(block_order = row_number(),
         y_min = y_order - 0.5 * b_w + (block_order - 1) * (b_w / 5),
         y_max = y_min + (b_w / 5)) %>% 
  inner_join(seq_17_146_kmer_df_rank %>% select(-kmer, -pep_len), 
             c("kmer_end", "kmer_start")) %>% 
  arrange(kmer_start, kmer_end)

# Process detected peptides data
seq_17_146_detected_df <- pep_nonc %>% 
  filter(protein_accession == "CONTRIB_GENCODE_c17norep146") %>% 
  inner_join(peptide, "peptide") %>% 
  dplyr::count(peptide, start_loc, end_loc) %>% 
  arrange(start_loc) %>% 
  mutate(pep_len = nchar(peptide),
         x_group = c(1,2,2,3,3,3,3,3)) %>% 
  group_by(x_group) %>% 
  mutate(y_order = row_number(),
         y_min = y_order - 0.5 * b_w,
         y_max = y_order + 0.5 * b_w) 

rectangles <- seq_17_146_detected_df %>% 
  group_by(x_group) %>% 
  summarise(min = min(start_loc), max = max(end_loc), .groups = "drop")

# Do not exclude kmers
seq_17_146_kmer_df_rank_all <- seq_17_146_kmer_df %>% 
  mutate(pep_len = nchar(kmer)) %>% group_by(pep_len) %>% 
  mutate(kmer_start = row_number(),
         kmer_end = row_number() + pep_len - 1) %>% ungroup() %>% 
  left_join(net_c17_146 %>% 
              inner_join(alleles_17_146, c("hla_type" = "hla_typing")), 
            c("kmer" = "peptide")) %>% 
  #filter(rank <= 2) %>% 
  group_by(kmer, pep_len, kmer_start, kmer_end) %>% 
  summarize(min_rank = min(rank), .groups = "drop")

seq_17_146_detected_df %>% inner_join(seq_17_146_kmer_df_rank_all, 
                                     c("start_loc" = "kmer_start", "end_loc" = "kmer_end"))

# Create plots
plot_17_146_pred <- ggplot() +
  geom_rect(data = rectangles,
            aes(xmin = min, xmax = max, ymin = 1-b_w, 
                ymax = max(kmer_17_146_p_data$y_max), group = x_group), 
            alpha = 0.2, fill = brewer.pal(5, "Set1")[3]) +
  geom_rect(data = kmer_17_146_p_data,
            aes(xmin = kmer_start, xmax = kmer_end, ymin = y_min, 
                ymax = y_max, fill = min_rank)) +
  black_theme + 
  scale_fill_gradient(name = "Minimum rank", low = grey_col[1], 
                      high = grey_col[7], limits = c(0, 2), 
                      breaks = c(0, 1, 2)) +
  scale_x_continuous(labels = NULL, limits = c(1, nchar(seq_17_146)), expand = c(0,0)) +
  scale_y_continuous(limits = c(1-b_w, NA), expand = c(0,0)) + 
  remove_y +
  theme(axis.line.y = element_blank(), axis.ticks.x = element_blank())

plot_17_146_det <- ggplot() +
  geom_rect(data = rectangles,
            aes(xmin = min, xmax = max, ymin = 1-b_w, 
                ymax = max(seq_17_146_detected_df$y_max), group = x_group), 
            alpha = 0.2, fill = brewer.pal(5, "Set1")[3]) +
  geom_rect(data = seq_17_146_detected_df,
            aes(xmin = start_loc, xmax = end_loc, ymin = y_min, 
                ymax = y_max, fill = n)) +
  black_theme +
  scale_x_continuous(limits = c(1, nchar(seq_17_146)), expand = c(0,0),
                     breaks = 1:nchar(seq_17_146),
                     labels = ~ ifelse(.x %% 10 == 0, .x, ""),
                     name = seq_17_146) +
  scale_fill_gradient(name = "Number of\ndetections\nin MS-runs", 
                      low = brewer.pal(9,"Greens")[6], 
                      high = brewer.pal(9,"Greens")[9], 
                      limits = c(0, NA)) +
  scale_y_reverse(limits = c(NA, 1-b_w)) + 
  remove_y +
  theme(axis.line.y = element_blank())

plot_17_146_shape <- ggplot(seq_17_146_contour) + 
  geom_step(aes(x = amino_acids, y = n), direction = "vh") +
  black_theme +
  scale_x_continuous(limits = c(1, nchar(seq_17_146)), expand = c(0,0), 
                     labels = NULL, name = NULL) +
  scale_y_continuous(name = NULL, expand = c(0,0), breaks = c(0,10,20)) +
  theme(axis.ticks.x = element_blank())

figure_6B <- plot_17_146_shape / plot_17_146_pred / plot_17_146_det

ggsave("Figure_6B.svg", device = "svg", path = plot_dir, figure_6B, 
       width = 2200, height = 1000, units = "px")
```

# Figure S1 - Atlas statistics numbers

```{r}
atlas_stats_hla <- read_excel(atlas_stats_file) %>% get_atlas_stats() 
atlas_stats_nonhla <- read_delim(atlas_stats_nonhla_file, delim = "\t") %>% 
  get_atlas_stats() 

atlas_color <- c(blue_col[1], grey_col[6])

max_pro <- max(atlas_stats_hla %>% filter(str_starts(name, "cum_pro")) %>% 
                 pull(value) %>% max(),
               atlas_stats_nonhla %>% filter(str_starts(name, "cum_pro")) %>% 
                 pull(value) %>% max())

create_atlas_plot <- function(plot_data, y_name, y_breaks, y_labels,
                              x_breaks = 0, x_labels = NULL, 
                              y_lim = NA) {
  count_label <- paste0(format(max(plot_data$value), big.mark = ","), 
                               " ", str_extract(y_name, "\\w+$"))
  c_label <- grobTree(textGrob(count_label, x = 0.5,  y = 1, just = c("center", "top"),
      gp = gpar(col = "black", fontsize = 9)))
  atlas_plot <- ggplot() +
    geom_ribbon(data = plot_data, 
                aes(x = spectra_searched, ymin = 0, ymax = value, 
                    group = name, fill = name), show.legend = F) +
    scale_fill_manual(values = atlas_color) +
    scale_x_continuous(name = "Cumulative number of MS/MS spectra identified",
                       breaks = x_breaks, labels = x_labels,
                       expand = c(0,0)) +
    scale_y_continuous(name = y_name, breaks = y_breaks, labels = y_labels,
                       expand = c(0,0), limits = c(0, y_lim)) +
    black_theme + annotation_custom(c_label)
  if (is.null(x_labels)) {atlas_plot <- atlas_plot + remove_x}
  return(atlas_plot)
}
  
p_hla_pep <- 
  create_atlas_plot(atlas_stats_hla %>% filter(str_starts(name, "cum_pep")),
                    "Number of distinct peptides", c(0, 2e5, 4e5, 6e5, 8e5),
                    c("0", "200k", "400k", "600k", "800k"))

p_nonhla_pep <- 
  create_atlas_plot(atlas_stats_nonhla %>% filter(str_starts(name, "cum_pep")),
                    "Number of distinct peptides", c(0, 1e6, 2e6, 3e6),
                    c("0", "1M", "2M", "3M"))

p_hla_pro <- 
  create_atlas_plot(atlas_stats_hla %>% filter(str_starts(name, "cum_pro")),
                    "Number of distinct proteins", c(0, 5e3, 10e3, 15e3),
                    c("0", "5k", "10k", "15k"), c(0, 1e8, 2e8, 3e8, 4e8, 5e8),
                    c("0", "100M", "200M", "300M", "400M", "500M"),
                    max_pro)

p_nonhla_pro <- 
  create_atlas_plot(atlas_stats_nonhla %>% filter(str_starts(name, "cum_pro")),
                    "Number of distinct proteins", c(0, 5e3, 10e3, 15e3),
                    c("0", "5k", "10k", "15k"), c(0, 1e8, 2e8, 3e8, 4e8, 5e8),
                    c("0", "100M", "200M", "300M", "400M", "500M"),
                    max_pro)

figure_S1 <- (p_hla_pep + p_nonhla_pep) / (p_hla_pro + p_nonhla_pro)

ggsave("Figure_S1AD.svg", device = "svg", path = plot_dir, figure_S1AD, 
       width = 2300, height = 1278, units = "px")
```

# Supplementary 2 - HLA class and ncORF peptide origin

## Figure S2A - Peptide classification

```{r}
barplot_data <- peptide_hla %>% group_by(peptide) %>% 
  mutate(hla_class = ifelse(n_distinct(hla_class) > 1, "Both", hla_class)) %>% 
  ungroup() %>% distinct(peptide, hla_class, pep_type) %>% 
  mutate(pep_len = nchar(peptide),
         pep_len_cat = ifelse(pep_len < 8, "7", 
                              ifelse(pep_len > 12, "13-35", "8-12"))) %>% 
  dplyr::count(hla_class, pep_type, pep_len_cat)

barplot_data_cat <- bind_rows(
  #Sample count
  peptide_hla %>% distinct(dataset_id, ms_name, hla_class) %>% 
    dplyr::count(hla_class) %>% mutate(cat = 1, pep_type = "all") %>% 
    dplyr::rename(group = hla_class),
  #HLA class all peptides
  barplot_data %>% group_by(hla_class) %>% summarise(n = sum(n)) %>% 
    mutate(cat = 2, pep_type = "all") %>% dplyr::rename(group = hla_class),
  #HLA class split peptides
  barplot_data %>% group_by(hla_class, pep_type) %>% summarise(n = sum(n)) %>% 
    mutate(cat = 3) %>% dplyr::rename(group = hla_class),
  #Length filtering
  barplot_data %>% filter((hla_class == "I" | hla_class == "Both"), 
                          pep_type != "other") %>% 
    group_by(pep_len_cat, pep_type) %>% summarise(n = sum(n)) %>% 
    mutate(cat = 4) %>% dplyr::rename(group = pep_len_cat),
  #Known HLA type filtering
  hla_df %>% distinct(peptide, pep_type) %>% dplyr::count(pep_type) %>% 
    mutate(cat = 5, group = "Known_HLA")) %>% 
  pivot_wider(names_from = c(group, cat), values_from = n) %>% 
  mutate(Unknown_HLA_5 = `8-12_4` - Known_HLA_5) %>% pivot_longer(-pep_type) %>% 
  filter(!is.na(value)) %>% 
  mutate(cat = as.integer(str_match(name, "\\d$")),
         group = as.character(str_match(name, ".*(?=_\\d$)"))) %>% 
  select(-name) %>% group_by(cat, pep_type) %>% 
  mutate(perc = value/sum(value), sums = sum(value)) %>% 
  filter(!(cat == 5 & pep_type == "other")) %>% 
  #add colors
  left_join(data.frame(group = c("I", "II", "Both", "8-12", "13-35",
                                      "Known_HLA", "Unknown_HLA"), 
                       fill_color = c(blue_col[1], grey_col[c(3,5)],
                                      blue_col[c(4, 2, 7, 5)])), 
            "group") %>% 
  mutate(cat = factor(cat, levels = c(1,2,3,4,5)),
         group = factor(group, c("I", "Both", "II", "7", "8-12", "13-35", 
                                 "Known_HLA", "Unknown_HLA")))

barplot_sums <- barplot_data_cat %>% distinct(cat, pep_type, sums) %>% 
  mutate(rowname = paste0(pep_type, "_", cat)) %>% column_to_rownames("rowname")

p_overview_1 <- ggplot() +
  geom_bar(data = barplot_data_cat %>% filter(cat == 1 | cat == 2),
           aes(x = cat, y = perc, fill = fill_color, group = group),
           stat = "identity") + black_theme +
  scale_fill_identity() + 
  scale_y_continuous(expand = c(0,0), name = element_blank()) +
  scale_x_discrete(name = element_blank(), 
                   labels = c(paste0(annotations %>% distinct(dataset_id) %>%
                                       nrow(), " datasets"), 
                              paste0("All peptides\n(n = ", 
                                     barplot_sums["all_2", "sums"],")")), 
                   expand = c(0,0)) +
  theme(axis.line = element_blank(), axis.ticks = element_blank(),
        axis.text.y = element_blank()) 

p_overview_2a <- ggplot() +
  geom_bar(data = barplot_data_cat %>% filter((cat == 3 | cat == 4 | cat == 5), 
                                              pep_type == "canonical"),
           aes(x = cat, y = value, fill = fill_color, group = group),
           stat = "identity") + black_theme +
  scale_fill_identity() +
  scale_y_continuous(name = element_blank(), expand = c(0,0)) +
  scale_x_discrete(name = element_blank(), expand = c(0,0),
                   labels = c(paste0("(n = ", 
                                     barplot_sums["canonical_3", "sums"], ")"),
                              paste0("(n = ", 
                                     barplot_sums["canonical_4", "sums"], ")"),
                              paste0("(n = ", 
                                     barplot_sums["canonical_5", "sums"], ")"))) +
  theme(axis.line = element_blank(), axis.ticks = element_blank(),
        axis.text.y = element_blank()) 

p_overview_2b <- ggplot() +
  geom_bar(data = barplot_data_cat %>% filter((cat == 3 | cat == 4 | cat == 5), 
                                              pep_type == "noncanonical"),
           aes(x = cat, y = value, fill = fill_color, group = group),
           stat = "identity") + black_theme +
  scale_fill_identity() +
  scale_y_continuous(name = element_blank(), expand = c(0,0)) +
  scale_x_discrete(name = element_blank(), expand = c(0,0),
                   labels = c(paste0("(n = ", 
                                     barplot_sums["noncanonical_3", "sums"], 
                                     ")\nHLA class on\nwhich peptide\nwas detected"),
                              paste0("(n = ", 
                                     barplot_sums["noncanonical_4", "sums"], 
                                     ")\nPeptide length"),
                              paste0("(n = ", 
                                     barplot_sums["noncanonical_5", "sums"], 
                                     ")\n HLA typing\nknown"))) +
  theme(axis.line = element_blank(), axis.ticks = element_blank(),
        axis.text.y = element_blank()) 

figure_S2A <- p_overview_1 + p_overview_2a + 
  plot_spacer() + p_overview_2b + 
  plot_layout(widths = c(2, 2))

ggsave("Figure_S2A.svg", device = "svg", path = plot_dir, figure_S2A,
       width = 2000, height = 1300, units = "px")
```

## Figure S2B - Classification on protein/ORF level

```{r}
pep_pro <- bind_rows(pep_can %>% semi_join(can_id, "protein_accession") %>% 
                       distinct(peptide, protein_accession),
                     pep_nonc %>% distinct(peptide, protein_accession))

protein_hla <- peptide_hla %>% distinct(peptide, hla_class) %>% 
  inner_join(pep_pro, "peptide") %>% distinct(protein_accession, hla_class) %>% 
  group_by(protein_accession) %>% 
  mutate(hla_class = ifelse(n() == 2, "Both", hla_class)) %>% ungroup() %>% 
  distinct(protein_accession, hla_class) %>% 
  mutate(is_can = !str_detect(protein_accession, "CONTRIB_GENCODE")) %>% 
  dplyr::count(is_can, hla_class) %>% 
  mutate(hla_class = factor(hla_class, levels = c("I", "Both", "II")))

figure_S2B1 <- ggplot(protein_hla %>% filter(is_can), 
                          aes(x = "Canonical\nproteins", y = n, 
                              fill = hla_class)) +
  geom_bar(stat = "identity", position = "stack", show.legend = F) + 
  black_theme +
  scale_x_discrete(expand = c(0,0), name = "") +
  scale_y_continuous(expand = c(0,0), name = "Number of canonical proteins",
                     breaks = c(0, 4e3, 8e3, 12e3,16e3), 
                     labels = c(0, "4k", "8k", "12k", "16k")) +
  scale_fill_manual(values = c(blue_col[1], grey_col[c(5,3)]))


figure_S2B2 <- ggplot(protein_hla %>% filter(!is_can), 
                          aes(x = "Non-canonical\nORFs", y = n, 
                              fill = hla_class)) +
  geom_bar(stat = "identity", position = "stack") + black_theme +
  scale_x_discrete(expand = c(0,0), name = "") +
  scale_y_continuous(expand = c(0,0), name = "Number of non-canonical ORFs",
                     breaks = c(0, 500, 1000, 1500), 
                     labels = c(0, "500", "1,000", "1,500")) +
  scale_fill_manual(values = c(blue_col[1], grey_col[c(5,3)]),
                    name = element_blank(),
                    labels = c("Detected only in\nHLA-I MS-runs",
                               "Detected in both\nHLA-I and HLA-II MS-runs",
                               "Detected only in\nHLA-II MS-runs"))

figure_S2B <- figure_S2B1 + figure_S2B2
```

## Figure S2C - Classification by peptide length

```{r}
pep_hla <- peptide %>% inner_join(annotations, c("dataset_id", "ms_name")) %>% 
  distinct(peptide, hla_class) %>% group_by(peptide) %>% 
  mutate(hla_class = ifelse(n() > 1, "Both", hla_class)) %>% ungroup() %>% 
  distinct(peptide, hla_class)

pep_class <- bind_rows(
  pep_nonc %>% distinct(peptide) %>% mutate(pep_type = "noncanonical"),
  pep_can %>% distinct(peptide) %>% mutate(pep_type = "canonical")) %>% 
  inner_join(pep_hla, "peptide") %>% 
  mutate(pep_len = ifelse(nchar(peptide) > 14, ">14 aa", "≤ 14aa")) %>% 
  dplyr::count(hla_class, pep_type, pep_len) %>% 
  group_by(pep_type) %>% mutate(perc = n / sum(n)*100,
                                hla_class = factor(hla_class, c("I", "Both", "II")))

figure_S2C <- ggplot(pep_class, aes(hla_class, perc, group = pep_len, fill = pep_len)) +
  geom_bar(stat = "identity") + black_theme +
  scale_x_discrete(expand = c(0,0), name = "HLA class on which peptide is detected") +
  scale_y_continuous(expand = c(0,0), name = "Percentage of all peptides") +
  scale_fill_manual(values = blue_col[c(6, 2)], name = "Peptide length") +
  facet_wrap(~pep_type)
```

## Figure S2D - Graph length distribution per dataset

```{r}
peptide_lendis <- peptide_hla %>% group_by(hla_class, dataset_id) %>% 
  mutate(p_count = n(), p_len = nchar(peptide)) %>% 
  group_by(dataset_id, hla_class, p_len) %>% 
  summarise(len_count = n(), p_count = first(p_count), 
            len_count_f = len_count/p_count) %>% mutate(dataset_hla = paste0(
            dataset_id, "__", hla_class))

lendis_n <- peptide_lendis %>% distinct(dataset_id, hla_class) %>% ungroup() %>% 
  dplyr::count(hla_class) %>% deframe()

figure_S2D <- ggplot(data = peptide_lendis %>% filter(p_len <= 30), 
       aes(x = p_len, y = len_count_f, group = dataset_hla, 
                                color = hla_class)) +
  geom_line(linewidth = 0.25) + geom_point(size = 0.5) + black_theme + 
  scale_y_continuous(expand = c(0, 0), limits = c(0,0.8), 
                     name = "Relative occurence") + 
  scale_x_continuous(expand = c(0, 0), limits = c(6.80,30), 
                     name = "Peptide length (aa)") +
  scale_color_manual(values = c(blue_col[1], grey_col[3]), name = "HLA class",
                     labels = c(paste0("HLA-I (n=", lendis_n["I"], 
                                       " dataset components)"),
                                paste0("HLA-II (n=", lendis_n["II"], 
                                       " dataset components)")))
```

## Figure S2E - Compare HLA I vs II

```{r}
orf_hlaclass <- peptide_hla %>%
  inner_join(pep_nonc, "peptide", relationship = "many-to-many") %>%
  distinct(peptide, hla_class, orf_biotype) %>%
  group_by(peptide, orf_biotype) %>%
  mutate(hla_class = ifelse(n() > 1, "II", hla_class)) %>%
  distinct() %>% ungroup() 

orf_hlaclass_count <- orf_hlaclass %>% dplyr::count(hla_class, orf_biotype) %>%
  group_by(hla_class) %>% mutate(maxi = max(n)) %>% ungroup() %>% 
  mutate(scale_factor = max(maxi) / min(maxi),
         orf_biotype = factor(orf_biotype, 
                              levels = unique(orf_biotype[order(hla_class, n)])),
         n_adj = ifelse(hla_class == "II", n * scale_factor, -n)) %>%
  arrange(hla_class, -n)
class_scale <- orf_hlaclass_count$scale_factor %>% first()

hla_n <- orf_hlaclass_count %>% dplyr::count(hla_class, wt = n) %>% deframe()

figure_S2E <- 
  ggplot(orf_hlaclass_count, aes(y = orf_biotype, x = n_adj, fill = hla_class)) + 
  geom_bar(stat = "identity", position = "identity", show.legend = F) + 
  black_theme +
  scale_x_continuous(position = "top", 
                     breaks = c(-1250,-1000,-750, -500, -250, 0, 20*class_scale, 
                                40*class_scale, 60*class_scale, 80*class_scale),
                     labels = c(1250, 1000, 750,500,250,0,20,40,60,80),
                     limits = c(-1250, 1250),
                     expand = c(0,0),
                     name = "# peptides of ORF biotype detected") + 
  scale_fill_manual(values = c(blue_col[1], grey_col[5]), 
                    labels = c(paste0("HLA-I (n=", hla_n["I"], " peptides)"),
                               paste0("HLA-II (n=", hla_n["II"], " peptides)"))) +
  theme(axis.line.y = element_blank(), axis.ticks.y = element_blank())
```

## Figure S2F - Check HLA-II peptides

```{r}
dataset_class <- annotations %>% distinct(dataset_id, hla_class) %>% 
  group_by(dataset_id) %>% 
  mutate(n_class = n(), 
         hla_class_dataset = ifelse(n_class == 2, "Both", hla_class)) %>% 
  distinct(dataset_id, hla_class_dataset) 

hlaii_orf_peptides_count <- peptide_hla %>%
  inner_join(orf_hlaclass %>% filter(hla_class == "II") %>% distinct(peptide),
             "peptide") %>% dplyr::count(peptide, hla_class, dataset_id) %>%
  inner_join(dataset_class, "dataset_id") %>%
  mutate(len = nchar(peptide)) %>% group_by(peptide, hla_class) %>%
  mutate(sum_n = sum(n)) %>% group_by(hla_class) %>%
  mutate(maxi = max(sum_n)) %>% ungroup() %>%
  mutate(scale_factor = max(maxi) / min(maxi),
         n = ifelse(hla_class == "II", n * scale_factor, -n),
         peptide = factor(peptide, levels =
                            rev(unique(peptide[order(hla_class,-sum_n)])))) %>% 
  group_by(peptide, hla_class, hla_class_dataset) %>% 
  mutate(color_num = row_number()) %>% ungroup %>% 
  mutate(color = case_when(
    hla_class_dataset == "I" & (color_num %% 2 == 0) ~ as.character(color_num+100),
    hla_class_dataset == "I" & (color_num %% 2 == 1) ~ as.character(color_num+100),
    hla_class_dataset == "II" & (color_num %% 2 == 0) ~ as.character(color_num+200),
    hla_class_dataset == "II" & (color_num %% 2 == 1) ~ as.character(color_num+200),
    color_num %% 2 == 0 ~ as.character(color_num+300),
    color_num %% 2 == 1 ~ as.character(color_num+300),
    T ~ NA
  ))

hla_color_mapping <- hlaii_orf_peptides_count %>% 
  distinct(color) %>% arrange(color) %>% 
  mutate(
    color_int = as.integer(color),
    color_hex = case_when(
      color_int < 200 & (color_int %% 2 == 0) ~ blue_col[1],
      color_int < 200 & (color_int %% 2 == 1) ~ blue_col[4],
      color_int < 300 & (color_int %% 2 == 0) ~ grey_col[1],
      color_int < 300 & (color_int %% 2 == 1) ~ grey_col[4],
      color_int %% 2 == 0 ~ pink_col[1],
      color_int %% 2 == 1 ~ pink_col[4],
      T ~ NA))

pep_len_mapping <- hlaii_orf_peptides_count %>% 
  arrange(peptide) %>% distinct(peptide, len)

hla_ii_scale <- hlaii_orf_peptides_count$scale_factor %>% unique()

detection_count <- hlaii_orf_peptides_count %>% 
  distinct(peptide, sum_n, hla_class) %>%
  dplyr::count(hla_class, wt = sum_n) %>% deframe()

figure_S2F <- ggplot(hlaii_orf_peptides_count, 
                      aes(y = peptide, x = n, fill = color)) + 
  geom_bar(stat = "identity", show.legend = F) + black_theme +
  scale_fill_manual(breaks = hla_color_mapping$color, 
                    values = hla_color_mapping$color_hex) +
  geom_vline(aes(xintercept = 0)) +
  scale_y_discrete(labels = pep_len_mapping$len, guide = guide_axis(n.dodge = 2),
                   name = paste0("Peptides (n = ", nrow(pep_len_mapping), ")")) + 
  scale_x_continuous(position = "top", 
                      breaks = c(-600, -400, -200, 0, 20*hla_ii_scale, 
                                 40*hla_ii_scale, 60*hla_ii_scale),
                      labels = c(600,400,200,0,20,40,60),
                      limits = c(-60*hla_ii_scale, 60*hla_ii_scale),
                     expand = c(0,0),
                     name = paste0("Number of peptide detections per dataset\n",
                                   "On HLA-I (n = ", detection_count["I"], ")\n",
                                   "On HLA-II (n = ", detection_count["II"], ")"))
```

## Figure S2B-F - Combine

```{r}
figure_S2BF <- figure_S2B + figure_S2C + figure_S2D + figure_S2E + free(figure_S2F)  + 
  plot_layout(design = "ABFFF\nCCFFF\nDDFFF\nEEFFF", 
              guides = 'collect')

ggsave("figure_S2BF.svg", device = "svg", path = plot_dir, figure_S2BF, 
       width = 3200, height = 2500, units = "px")
```

# Figure S3 - ORF statistics

## Figure S3A-D - ORF overview

```{r}
biotype_counts <- orf_pep_stats %>% mutate(detected = n_pep >= 1) %>% 
  dplyr::count(orf_biotype, detected) %>% 
  pivot_wider(names_from = detected, values_from = n) %>% 
  mutate(total = `TRUE` + `FALSE`,
         perc = `TRUE` / total,
         mean_perc = sum(`TRUE`) / sum(total)) %>% rowwise() %>% 
  mutate(p_val = binom.test(`TRUE`, total, p = mean_perc)$p.value * nrow(.))

orf_pep_stats_len <- orf_pep_stats %>% 
  mutate(is_det = ifelse(n_pep > 0, T, F)) %>% 
  mutate(len_cat = cut(orf_len, breaks = c(15, 30, 50, 75, 100, Inf), 
                       labels = c("16-30","31-50","51-75","76-100",">100"))) %>% 
  dplyr::count(len_cat, is_det) %>% group_by(len_cat) %>% 
  mutate(total = sum(n)) %>% ungroup() %>% rowwise() %>% 
  mutate(perc = n/total, conf_low = binom.test(n,total)$conf.int[1],
         conf_high = binom.test(n,total)$conf.int[2],
         conf_low = ifelse(is_det, conf_low, NA),
         conf_high = ifelse(is_det, conf_high, NA))

orf_pep_stats_ref <- biotype_counts %>% 
  pivot_longer(c(`TRUE`, `FALSE`)) %>% 
  dplyr::rename(n = value, is_det = name) %>% rowwise() %>% 
  mutate(is_det = is_det == "TRUE",
         perc = n/total,
         conf_low = binom.test(n,total)$conf.int[1],
         conf_high = binom.test(n,total)$conf.int[2],
         conf_low = ifelse(is_det, conf_low, NA),
         conf_high = ifelse(is_det, conf_high, NA))

orf_len_enrichment <- orf_pep_stats_len %>% filter(is_det) %>% 
  mutate(cat = ifelse(len_cat == "16-30", "low", "high")) %>% 
  group_by(cat) %>% summarise(perc = sum(n)/sum(total)*100) %>% 
  pivot_wider(names_from = cat, values_from = perc) %>% 
  mutate(enrichment = (high - low)/high)

figure_S3A <- ggplot(orf_pep_stats_ref, aes(fill = is_det, y = n, x = orf_biotype)) + 
  geom_bar(position = "stack", stat = "identity", show.legend = F) + black_theme + 
  scale_x_discrete(name = element_blank(), expand = c(0,0)) +
  scale_y_continuous(name = "Number of non-canonical ORFs", expand = c(0,0)) +
  scale_fill_manual(values = green_col[c(4,1)]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_S3B <- ggplot(orf_pep_stats_ref, 
                     aes(fill = is_det, y = perc, x = orf_biotype)) + 
  geom_bar(position = "stack", stat = "identity", show.legend = F) + black_theme +
  scale_x_discrete(name = element_blank(), expand = c(0,0)) +
  scale_y_continuous(name = "Percentage of non-canonical ORFs", 
                     expand = c(0,0), breaks = c(0, 0.25, 0.5, 0.75, 1),
                     labels = c("0%", "25%", "50%", "75%", "100%")) +
  scale_fill_manual(values = green_col[c(4,1)]) +
  geom_linerange(aes(ymin = conf_low, ymax=  conf_high)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_S3C <- ggplot(orf_pep_stats_len, aes(fill = is_det, y = n, x = len_cat)) + 
  geom_bar(position = "stack", stat = "identity", show.legend = F) + black_theme + 
  scale_x_discrete(name = "Orf length (aa)", expand = c(0,0)) +
  scale_y_continuous(name = "Number of non-canonical ORFs", expand = c(0,0)) +
  scale_fill_manual(values = green_col[c(4,1)]) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_S3D <- ggplot(orf_pep_stats_len, aes(fill = is_det, y = perc, x = len_cat)) + 
  geom_bar(position = "stack", stat = "identity", show.legend = F) + black_theme +
  scale_x_discrete(name = "Orf length (aa)", expand = c(0,0)) +
  scale_y_continuous(name = "Percentage of non-canonical ORFs", expand = c(0,0),
                     breaks = c(0, 0.25, 0.5, 0.75, 1),
                     labels = c("0%", "25%", "50%", "75%", "100%")) +
  scale_fill_manual(values = green_col[c(4,1)]) +
  geom_linerange(aes(ymin = conf_low, ymax=  conf_high)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

figure_S3 <- figure_S3A + figure_S3B + figure_S3C + figure_S3D +
  plot_layout(nrow = 1)

ggsave("Figure_S3AD.svg", device = "svg", path = plot_dir, figure_S3,
       width = 2340, height = 1000, units = "px")
```

## Figure S3E - Compare cancer vs non-cancer with cancer genes

```{r}
cancer_census <- read_csv(cancer_census_file)
colnames(cancer_census) <- str_replace_all(colnames(cancer_census), " ", "_")

orf_gene <- orf_sequences %>% select(-protein) %>% left_join(
  cancer_census %>% distinct(Gene_Symbol, Role_in_Cancer) %>% mutate(is_cancergene = T),
  c("gene_name" = "Gene_Symbol"))

orf_cancer <- peptide_hla %>%
  inner_join(pep_nonc, "peptide", relationship = "many-to-many") %>% 
  left_join(orf_gene, c("orf_name", "orf_biotype")) %>% 
  distinct(peptide, is_cancer, orf_biotype, is_cancergene) %>% 
  group_by(peptide, orf_biotype) %>%
  mutate(is_cancer = ifelse(n() > 1, "Two", is_cancer),
         is_cancergene = ifelse(is.na(is_cancergene), F, T)) %>%
  distinct() %>% ungroup() 

orf_cancer_count <- orf_cancer %>% 
  dplyr::count(is_cancer, orf_biotype, is_cancergene) %>%
  group_by(is_cancer) %>% 
  mutate(orf_biotype = str_replace_all(orf_biotype, " ", "\n"),
         total = sum(n), perc = n/total*100,
         orf_biotype = factor(orf_biotype, levels = 
                                unique(orf_biotype[order(is_cancer, -perc)])),
         is_cancer = ifelse(is_cancer == "Two", "Both", is_cancer),
         is_cancer = factor(is_cancer, levels = 
                              c("Cancer", "Both", "Non-cancer"))) %>% 
  group_by(is_cancer, orf_biotype) %>% 
  mutate(sumn = sum(n)) %>% rowwise() %>% 
  mutate(conf_low = ifelse(!is_cancergene, 
                           binom.test(sumn,total)$conf.int[1]*100, NA),
         conf_high = ifelse(!is_cancergene, 
                            binom.test(sumn,total)$conf.int[2]*100, NA),
         color = case_when(
           is_cancergene ~ red_col,
           is_cancer == "Cancer" ~ blue_col[2],
           is_cancer == "Non-cancer" ~ blue_col[6],
           T ~ grey_col[4]),
         color = factor(color, levels = c(blue_col[2], blue_col[6], 
                                          grey_col[4], red_col)),
         x_axis = paste0(orf_biotype, is_cancer)) %>% 
  arrange(orf_biotype, is_cancer) %>% 
  mutate(x_axis = factor(x_axis, levels = unique(x_axis)))

cancer_x_labels <- orf_cancer_count %>% filter(is_cancer == "Both") %>% 
  arrange(orf_biotype) %>% distinct(x_axis, orf_biotype)

cancer_counts <- orf_cancer_count %>% ungroup() %>% 
  distinct(is_cancer, total) %>% deframe()

cancergene_count <- orf_cancer_count %>% filter(is_cancergene) %>% 
  pull(n) %>% sum()

cancer_colors <- c(blue_col[2], grey_col[4], blue_col[6])

figure_S3E <- 
  ggplot(orf_cancer_count, aes(x = x_axis, y = perc, fill = color)) + 
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_identity(guide = "legend", breaks = c(cancer_colors, red_col),
                      labels = c(paste0("Cancer\n(n = ", 
                                        cancer_counts["Cancer"], " peptides)"), 
                                 paste0("Both\n(n = ", 
                                        cancer_counts["Both"], " peptides)"),
                                 paste0("Non-cancer\n(n = ", 
                                        cancer_counts["Non-cancer"], 
                                        " peptides)"),
                                 paste0("Peptide from ORF\nlocated at cancergene\n(n = ", 
                                        cancergene_count, " peptides)")),
                      name = "Sample type in which\npeptide was detected") + 
  black_theme +
  scale_y_continuous(expand = c(0, 0), name = "% of ORF biotype detected",
                     breaks = c(0,10,20,30,40, 50),
                     labels = c("0%", "10%", "20%", "30%", "40%", "50%")) +
  scale_x_discrete(expand = c(0, 0), name = element_blank(),
                   breaks = cancer_x_labels$x_axis, 
                   labels = cancer_x_labels$orf_biotype) +
  facet_grid( ~ orf_biotype , scales = "free") +
  geom_linerange(aes(ymin = conf_low, ymax=  conf_high)) +
     theme(axis.text.x = element_text(angle = 60,  hjust = 1),
           strip.text.x = element_blank())

ggsave("Figure_S3E.svg", device = "svg", path = plot_dir, figure_S3E,
       width = 1400, height = 1200, units = "px")
```

## Figure S3F - Compare on peptide level

```{r}
neworf_cancer <- orf_cancer %>% distinct(peptide, is_cancer) %>% 
  mutate(len = nchar(peptide),
         hydro = hydrophobicity(peptide, "KyteDoolittle"),
         weight = mw(peptide)/len,
         pi = pI(peptide),
         chrg = charge(peptide), 
         is_cancer = ifelse(is_cancer == "Two", "Both", is_cancer),
         is_cancer = factor(is_cancer, 
                            levels = c("Cancer", "Both", "Non-cancer")))

len_dist_can <- create_dis_plot(ggplot(neworf_cancer, 
                                       aes(x = len, color = is_cancer)), 
                "Peptide length", cancer_colors)

hydro_dist_can <- create_dis_plot(ggplot(neworf_cancer, 
                                     aes(x = hydro, color = is_cancer)), 
                "Hydrophobicity (Kyte-Doolittle)", cancer_colors)

weight_dist_can <- create_dis_plot(ggplot(neworf_cancer, 
                                      aes(x = weight, color = is_cancer)), 
                "Mean mass per amino acid (u)", cancer_colors, 
                y_name = element_blank())

pi_dist_can <- create_dis_plot(ggplot(neworf_cancer, 
                                      aes(x = pi, color = is_cancer)), 
                "Isoelectric point", cancer_colors, y_name = element_blank())

figure_S3F <- wrap_plots(list(len_dist_can, weight_dist_can, 
                              hydro_dist_can, pi_dist_can), 
                         ncol = 2, nrow = 2)

ggsave("Figure_S3F.svg", device = "svg", path = plot_dir, figure_S3F,
       width = 1200, height = 1200, units = "px")
```

# Figure S4 - Binding predictions

## Figure S4AB - 9mers per hla-typing

### Extract 9mers from non-canonical ORFs

```{r}
k <- 9
nonc_seq <- can_nonc_seq %>% 
              filter(str_starts(protein_accession, "CONTRIB_GENCODE_c"))

k_mers_nonc <- apply(nonc_seq, 1, function(x) extract_kmers(x["protein"], k))
names(k_mers_nonc) <- paste0(nonc_seq$protein_accession, "__")
kmer_df_nonc <- k_mers_nonc %>% unlist() %>% enframe() %>% 
  mutate(protein_accession = str_extract(name, ".*(?=__)")) %>%
  group_by(protein_accession) %>% 
  mutate(kmer_loc = row_number(),
         end_offset = max(kmer_loc) - kmer_loc) %>% 
  select(-name) %>% dplyr::rename(kmer = value) %>% ungroup() %>% 
  left_join(pep_net_orf %>% filter(is_orf) %>% distinct(peptide) %>% 
              mutate(kmer_detected = T), c("kmer" = "peptide")) %>% 
  left_join(pep_nonc_long %>% ungroup() %>% distinct(protein_accession) %>% 
              mutate(orf_detected = T), "protein_accession") %>% 
  replace_na(list(kmer_detected = F, orf_detected = F))
```

### Run NetMHCpan and process data (only once)

```{r}
nonc_kmers <- kmer_df_nonc %>% distinct(kmer) %>% rename("peptide" = 1)

write_net("", nonc_kmers)

found_alleles <- hla_df %>% distinct(hla_type) %>% 
  arrange(hla_type) %>% mutate(
  hla_type = paste0("HLA-", str_extract(hla_type, "^[A-Z]\\d\\d"), ":", 
                  str_extract(hla_type, "(?<=^[A-Z]\\d\\d)\\d+"))
)

write.table(found_alleles$hla_type, paste0(net_input, "/net_alleles.txt"), 
            row.names = F, col.names = F, quote = F)

# Run netMHCpan!

net_files <- list.files(path = file.path(processed_dir, "netmhcpan_allkmers9"), 
  pattern = paste0("net_.*.xls"), full.names = TRUE)

save(net_files, nonc_kmers, file = file.path(processed_dir, "retrieve_net_kmers_data.Rds"))

# Run retrieve_netmhcpan_kmers_nonc.R
```

### Create figure S4A

```{r}
net_kmers_nonc <- readRDS(file.path(processed_dir, "net_kmers_nonc.Rds"))

ms_alleles <- hla_df %>% distinct(peptide, hla_typing, hla_type) %>% 
  inner_join(pep_nonc_long %>% distinct(peptide, protein_accession), 
             "peptide") %>% 
  distinct(protein_accession, hla_typing, hla_type) %>% 
  inner_join(kmer_df_nonc, "protein_accession") %>% 
  inner_join(net_kmers_nonc, c("kmer" = "peptide", "hla_type")) %>% 
  left_join(hla_df %>% distinct(peptide, hla_typing, hla_type) %>% 
               mutate(kmer_hla_detected = T),
            c("kmer" = "peptide", "hla_type", "hla_typing")) %>% 
  replace_na(list(kmer_hla_detected = F)) %>% 
  distinct(hla_type, rank, kmer_hla_detected, hla_typing, kmer) %>% 
  group_by(kmer_hla_detected, hla_typing, kmer) %>% 
  summarise(min_rank = min(rank)) %>% ungroup() %>% 
  mutate(detected = ifelse(kmer_hla_detected, "Detected", "Undetected"),
         kmer_hla_detected = factor(kmer_hla_detected, levels = c(T,F)))

min_rank_det_pval <- perform_wtest(ms_alleles %>% 
                                     dplyr::rename(y_val = min_rank), 1)
p_label_min_rank_det <- grobTree(textGrob(min_rank_det_pval, x = 0.5,  
                                          y = 1, just = c("center", "top"),
  gp = gpar(col = "black", fontsize = 7)))

figure_S4A <- ggplot(ms_alleles, 
                         aes(kmer_hla_detected, y = min_rank, 
                             fill = kmer_hla_detected)) +
  geom_violin(trim = T, show.legend = F, scale = "width") + 
  geom_boxplot(fill = "transparent", outlier.shape = NA) + black_theme +
  scale_fill_manual(values = green_col[c(2,5)]) +
  scale_x_discrete(name = element_blank(), breaks = c(T,F),
                   labels = c("Detected\n9mers", "Undetected\n9mers")) +
  scale_y_continuous(trans = scales::pseudo_log_trans(base = 10), limits = c(0,100),
                     breaks = c(0, 2, 25, 50, 75, 100),
                     name = "Minimum rank of 9mer\nfrom detected ncORFs\nper HLA-typing") +
  geom_hline(yintercept = 2, linetype = "dashed") + 
  annotation_custom(p_label_min_rank_det)
```

### Create figure S4B

```{r}
nonc_alleles <- hla_df %>% semi_join(pep_nonc, "peptide") %>% 
  distinct(hla_type)

nonc_kmer_all_allele_perc <- kmer_df_nonc %>% 
  inner_join(net_kmers_nonc, c("kmer" = "peptide")) %>% 
  semi_join(nonc_alleles, "hla_type") %>% 
  distinct(kmer, protein_accession, orf_detected, rank, hla_type)  %>% 
  group_by(kmer, protein_accession, orf_detected) %>% 
  summarise(min_rank = min(rank)) %>% 
  mutate(detected = ifelse(orf_detected, "Detected", "Undetected"),
         orf_detected = factor(orf_detected, levels = c(T,F))) %>% 
  mutate(is_binding = min_rank <= 2) %>% ungroup() %>% 
  dplyr::count(protein_accession, orf_detected, is_binding, detected) %>% 
  pivot_wider(names_from = is_binding, values_from = n, values_fill = 0) %>% 
  mutate(total = `TRUE` + `FALSE`, perc = `TRUE`/total*100)

perc_rank_orf_pval <- perform_wtest(nonc_kmer_all_allele_perc %>% 
                                     dplyr::rename(y_val = perc), 1)
p_label_perc_rank_orf <- grobTree(textGrob(perc_rank_orf_pval, x = 0.5,  y = 1, 
  just = c("center", "top"), gp = gpar(col = "black", fontsize = 7)))

figure_S4B <- ggplot(nonc_kmer_all_allele_perc, 
       aes(orf_detected, y = perc, fill = orf_detected)) +
    geom_violin(trim = T, show.legend = F) + 
    geom_boxplot(fill = "transparent", outlier.shape = NA) + black_theme +
    scale_fill_manual(values = green_col[c(2,5)]) +
    scale_x_discrete(name = element_blank(), breaks = c(T,F),
                     labels = c("9mers from\ndetected\nncORFs", 
                                "9mers from\nundetected\nncORFs")) +
  scale_y_continuous(name = paste0("% 9mers with rank ≤ 2 across\n", 
                                   nrow(nonc_alleles), " alleles per ncORF"), 
                     limits = c(0,100),
                     breaks = c(0, 25, 50, 75, 100),
                     labels = c("0%", "25%", "50%", "75%", "100%")) + 
  annotation_custom(p_label_perc_rank_orf)
```

### Figure S4A-B - Combine

```{r}
figure_S4AB <- figure_S4A + figure_S4B

ggsave("Figure_S4AB.svg", device = "svg", path = plot_dir, figure_S3,
       width = 1100, height = 1000, units = "px")
```

## Figure S4C - Hydrophobicity

```{r}
h_lab <- "Hydrophobicity\n(Kyte-Doolittle)"
h_cats <- c("Non-canonical__start", "Non-canonical__end")

figure_S4C_plots <- lapply(1:2, function(i) {
  cat <- h_cats[i]
  plot_data <- seq_stats_properties %>% filter(pro_cat == cat) %>% 
    mutate(y_val = hydro)
  p_val <- perform_wtest(plot_data, 8)
  create_violin(plot_data, h_lab, p_val, c(-4.1, 2.5), 
                n_colors = c((4*i-3):(4*i-2)))
})

figure_S4C_plots[[2]] <- figure_S4C_plots[[2]] + remove_y
figure_S4C <- wrap_plots(figure_S4C_plots, ncol = 2)

ggsave("Figure_4A.svg", device = "svg", path = plot_dir, figure_4A, 
       width = 1100, height = 800, units = "px")
```

## Figure S4D - Hydrophobicity per biotype

```{r}
seq_stats_hydro_biotype <- seq_stats_hydro %>% 
  group_by(orf_biotype, c_offset, detected) %>% 
  mutate(low_conf = t.test(hydro)$conf.int[1], 
         high_conf = t.test(hydro)$conf.int[2]) %>% 
  summarise(hydro = mean(hydro), low_conf = first(low_conf),
            high_conf = first(high_conf)) %>% 
  mutate(orf_biotype = factor(orf_biotype, 
                          levels = c("CDS", "uORFs", "uoORFs", "intORFs", 
                                     "doORFs", "dORFs", "lncRNA ORFs", 
                                     "Processed transcript ORFs")))

figure_S4D <- ggplot(seq_stats_hydro_biotype, 
       aes(x = c_offset, y = hydro, group = detected)) + 
  geom_linerange(aes(ymin = low_conf, ymax=  high_conf, color = detected), 
                 alpha = 0.5, linewidth = 0.3) +
  geom_point(mapping = aes(color = detected), size = .2) + 
  geom_smooth(method = "loess", formula = y ~ x, se = F, 
              mapping = aes(color = detected), linewidth = .7)  +
  scale_x_continuous(expand = c(0, 0), 
                     name = "Position relative to C-terminus") +
  scale_y_continuous(limits = c(-3.8, 2.6), expand = c(0, 0),
                     name = "Hydrophobicity\n(Kyte-Doolittle)") +
  scale_color_manual(values = c(green_col[1], blue_col[2])) +
  black_theme +
  facet_wrap(~orf_biotype, ncol = 4)

ggsave("figure_S4D.svg", device = "svg", path = plot_dir, figure_S4D,
       width = 2516.16, height = 900, units = "px") 
```

## Figure S4E - Expression per biotype

```{r}
orf_gtex_split <- orf_gtex %>% make_biotype_levels %>% split(~orf_biotype)

gtex_violins <- lapply(names(orf_gtex_split), function(i) {
  p_val <- perform_wtest(orf_gtex_split[[i]], 7)
  expression_violin <- create_violin(orf_gtex_split[[i]], 
    "Mean FPKM of gene\nexpressing ncORF in GTEx", p_val, c(0, 1600), 
    n_colors = c(3,4), pseudo_log = T, x_label = i, 
    y_breaks = c(0,10,250,500,1000,2000))
})

gtex_violins_f <- lapply(1:7, function(i) {
  one_plot <- gtex_violins[[i]]
  if (i > 1) {
    one_plot <- one_plot + remove_y
  }
  return(one_plot)
})

figure_S4E <- wrap_plots(gtex_violins_f, nrow = 1)

ggsave("figure_S4E.svg", device = "svg", path = plot_dir, figure_S4E, 
      width = 2300, height = 600, units = "px")
```

## Figure S4F - Expression per tissue

```{r}
gtex_tissue <- gtex_data %>% 
  dplyr::rename(gene_id = X) %>% pivot_longer(-gene_id) %>% 
  dplyr::rename(tissue_raw = name, fpkm = value) %>% 
  mutate(tissue = str_replace_all(str_extract(tissue_raw, "^\\w+"), "_", " "),
         tissue = case_when(
           tissue == "Adipose" ~ "Adipose Tissue",
           tissue == "Artery" ~ "Blood Vessel",
           tissue_raw == "Cells.EBV_transf.lymphocytes" ~ "Blood",
           tissue_raw == "Cells.transformed_fibroblasts" ~ "Skin",
           tissue == "FallopianTube" ~ "Fallopian Tube",
           tissue == "Adipose" ~ "Adipose Tissue",
           tissue == "Minor Salivary Gland" ~ "Salivary Gland",
           tissue == "Whole Blood" ~ "Blood",
           T ~ tissue)) %>% 
  group_by(tissue, gene_id) %>% summarise(fpkm = mean(fpkm))

orf_gtex_tissue <- seq_stats %>% 
  inner_join(orf_sequences %>% 
               select(protein_accession, gene_id), "protein_accession") %>% 
  inner_join(gtex_tissue, "gene_id")

# Statistical test per tissue
p_tissue <- orf_gtex_tissue %>% mutate(y_val = fpkm) %>% split(~ tissue) %>% 
  sapply(function(x) perform_wtest(x, 29, raw = T))
max(p_tissue)

figure_S4F <- ggplot(orf_gtex_tissue, aes(detected, fpkm, color = detected)) +
  geom_boxplot(outliers = F) + black_theme +
  scale_y_continuous(trans = scales::pseudo_log_trans(base = 10), 
                     breaks = c(0, 2, 10, 50, 250),
                     name = "Mean FPKM of gene\nexpressing ncORF in GTEx") +
  scale_x_discrete(labels = c(), name = element_blank()) + 
  facet_wrap(~ tissue, nrow = 1) +
  scale_color_manual(values = green_col[c(1,4)])

ggsave("Figure_S4F.svg", device = "svg", path = plot_dir, figure_S4F, 
      width = 2300, height = 600, units = "px")
```

## Figure S4G - Ligand Atlas quality

```{r}
figure_S4G <- ggplot(peplen_plot_data %>% filter(dataset_id == "PXD019643") %>% 
                       arrange(orf_count), 
                     aes(x = avg_len, y = perc, color = orf_count)) + 
  geom_point(aes(size = pep_count), stroke = 0) + black_theme + 
  theme(legend.direction = "vertical", legend.box = "horizontal") +
  scale_color_gradient(name = "# non-can\npeptides",
                      low = grey_col[1], high = red_col,
                      limits = c(0, 45),
                      breaks = c(0,10,20,30,40)) +
  scale_size_continuous(breaks = c(2000, 4000, 6000), 
                        labels = c("2,000", "4,000", "6,000"),
                        name = "# total\npeptides") +
  scale_y_continuous(expand = c(0, 0), limits = c(0,1.01), 
                     name = "% binders (rank ≤ 2)\namongst peptides from 8-12 aa", 
                     breaks = c(0, 0.25, 0.5, 0.75, 1),
                     labels = c("0%", "25%", "50%", "75%", "100%")) +
  scale_x_continuous(expand = c(0, 0), limits = c(8.7,15), 
                     name = "Mean peptide length",
                     breaks = c(10, 12, 14)) 

ggsave("figure_S4G.svg", device = "svg", path = plot_dir, figure_S4G, 
       width = 1237.205, height = 600, units = "px")
```

## Figure S4H - LigandAtlas - GTEX comparison

```{r}
ligand_atlas_genes <- pep_net_orf %>% ungroup() %>% 
  filter(dataset_id == "PXD019643", is_binding, pep_type == "noncanonical") %>% 
  distinct(peptide, pep_type) %>% 
  inner_join(pep_nonc, "peptide") 

gtex_df <- gtex_data %>% 
  dplyr::rename(gene_id = X) %>% pivot_longer(-gene_id) %>% 
  arrange(name) %>%  mutate(ligand_tissue = case_when(
    name == "Kidney.Cortex" ~ "Kidney",
    name == "Stomach" ~ "Stomach",
    name == "Pancreas" ~ "Pancreas",
    str_detect(name, "Esophagus") ~ "Esophagus",
    name == "Muscle.Skeletal" ~ "Muscle",
    name == "Prostate" ~ "Prostate",
    name == "Spleen" ~ "Spleen",
    name == "Brain.Cerebellum" ~ "Cerebellum",
    str_detect(name, "Cortex") ~ "Brain (Cerebral cortex)",
    name == "Adrenal_Gland" ~ "Adrenal Gland",
    str_detect(name, "Colon") ~ "Colon",
    str_detect(name, "Heart") ~ "Heart",
    name == "Lung" ~ "Lung",
    name == "Breast.Mammary_tissue" ~ "Mamma (Breast)",
    str_detect(name, "Skin") ~ "Skin",
    name == "Bladder" ~ "Bladder",
    name == "Artery.Aorta" ~ "Aorta",
    name == "Liver" ~ "Liver",
    name == "Thyroid" ~ "Thyroid",
    name == "Small_Intestine" ~ "Small Intestine",
    name == "Ovary" ~ "Ovary",
    name == "Uterus" ~ "Uterus",
    name == "Brain.Cerebellar_Hemisphere" ~ "Cerebellum")) %>% 
  filter(!is.na(ligand_tissue)) %>% 
  mutate(gtex_name = 
           str_replace_all(name, "_", " ") %>% 
           str_replace("[.]+", " - ") %>% 
           str_replace_all("[.]", "") %>% 
           str_replace("FrontalC", "Frontal C") %>% 
           str_replace("Gastroesophjunction", "Gastroesophegeal Junction") %>% 
           str_replace("LeftV", "Left V") %>% 
           str_replace("suprapubic", "Suprapubic") %>% 
           str_replace("lower leg", "Lower Leg") %>% 
           str_replace("appendage", "Appendage")) %>% 
  semi_join(ligand_atlas_genes, "gene_id") %>% 
  mutate(ligand_tissue = factor(ligand_tissue, 
                                levels = ligand_atlas_data$tissue_type)) %>% 
  arrange(ligand_tissue) %>% 
  mutate(gtex_name = factor(gtex_name, levels = unique(gtex_name)))

  mjfigure_S4H <- ggplot() +
  geom_jitter(gtex_df %>% filter(value <= 50), mapping = aes(gtex_name, value), 
              stat = "identity", size = .01) + black_theme +
  geom_boxplot(gtex_df, mapping = aes(gtex_name, value), 
               alpha = 0.8, outlier.shape = NA, size = .1) +
  coord_cartesian(ylim = c(0,50)) +
  scale_x_discrete(expand = c(0,0), name = element_blank()) + 
  scale_y_continuous(expand = c(0,0),
                     name = "Mean TPM of gene") +
  scale_fill_manual(values = green_col[4]) +
  theme(axis.text.x = element_text(angle = 60,  hjust = 1))

ggsave("Figure_S4F.svg", device = "svg", path = plot_dir, figure_S4H,
        width = 1300, height = 800, units = "px")
```

# Figure S5 - HLA c5norep142

## Create c5norep142 kmers

```{r}
seq_5_142 <- can_nonc_seq %>% 
  filter(protein_accession == "CONTRIB_GENCODE_c5norep142") %>% 
  pull(protein)

seq_5_142_kmer_df<- sapply(8:12, function(i) extract_kmers(seq_5_142, i)) %>% 
  unlist() %>% enframe() %>% dplyr::rename(kmer = value) %>% select(-name)
```

## Run NetMHCpan (only once)

```{r}
unique_kmers <- seq_5_142_kmer_df %>% distinct(kmer) %>% rename("peptide" = 1)
write_net("", unique_kmers)

found_alleles <- hla_df %>% select(hla_type) %>% distinct() %>% 
  arrange(hla_type) %>% mutate(
  hla_type = paste0("HLA-", str_extract(hla_type, "^[A-Z]\\d\\d"), ":", 
                  str_extract(hla_type, "(?<=^[A-Z]\\d\\d)\\d+"))
)

write.table(found_alleles$hla_type, paste0(net_input, "/net_alleles.txt"), 
            row.names = F, col.names = F, quote = F)

# Run NetMHCpan
```
## Read NetMHCpan results
### Either read prediction results
```{r}
# Read and process netmhcpan files
net_c5_142 <- list.files(path = file.path(processed_dir, 
                                          "netmhcpan_kmers_c5norep142"), 
                         pattern = "net_.*.xls", full.names = TRUE) %>%
  map_df(read_netmhcpan)
```
### Or use pre-generated data (from github)
```{r}
net_c5_142 <- readRDS(file.path(raw_dir, net_c5_142.RDS)
```

## Create Figure S6
```{r}

# Process alleles
alleles_5_142 <- pep_nonc %>% 
  filter(protein_accession == "CONTRIB_GENCODE_c5norep142") %>% 
  inner_join(hla_df, "peptide") %>% 
  separate_rows(hla_typing, sep = ",") %>% 
  dplyr::count(hla_typing)

# Process kmer data
seq_5_142_kmer_df_rank <- seq_5_142_kmer_df %>% 
  mutate(pep_len = nchar(kmer)) %>% group_by(pep_len) %>% 
  mutate(kmer_start = row_number(),
         kmer_end = row_number() + pep_len - 1) %>% ungroup() %>% 
  left_join(net_c5_142 %>% 
              inner_join(alleles_5_142, c("hla_type" = "hla_typing")), 
            c("kmer" = "peptide")) %>% 
  filter(rank <= 2) %>% 
  group_by(kmer, pep_len, kmer_start, kmer_end) %>% 
  summarize(min_rank = min(rank), .groups = "drop")

# Process contour data
seq_5_142_contour <- seq_5_142_kmer_df_rank %>%
  rowwise() %>% 
  mutate(amino_acids = list(kmer_start:kmer_end)) %>%
  unnest(amino_acids) %>%
  dplyr::count(amino_acids) %>% 
  right_join(tibble(amino_acids = 1:100), by = "amino_acids") %>% 
  replace_na(list(n = 0)) %>% 
  mutate(amino_acids = amino_acids + 0.5) %>% 
  add_row(amino_acids = 1, n = 0, .before = 1) %>% 
  mutate(n = ifelse(amino_acids == 1, lead(n), n),
         amino_acids = ifelse(amino_acids == 100.5, 100, amino_acids))

# Process kmer plot data
kmer_5_142_p_data <- seq_5_142_kmer_df_rank %>% 
  group_by(kmer_end) %>%
  summarise(min_start = min(kmer_start),
            kmer_start = list(kmer_start),
            pep_num = n()) %>%
  mutate(x_group = rep(1:6, c(8, 5, 10, 8, 8, 5))) %>%
  group_by(x_group) %>%
  mutate(y_order = row_number()) %>% 
  unnest(kmer_start) %>% 
  
  mutate(pep_len = factor(kmer_end - kmer_start + 1, levels = 12:8)) %>% 
  group_by(kmer_end) %>% 
  arrange(pep_len) %>% 
  mutate(block_order = row_number(),
         y_min = y_order - 0.5 * b_w + (block_order - 1) * (b_w / 5),
         y_max = y_min + (b_w / 5)) %>% 
  inner_join(seq_5_142_kmer_df_rank %>% select(-kmer, -pep_len), 
             c("kmer_end", "kmer_start")) %>% 
  arrange(kmer_start, kmer_end)

# Process detected peptides data
seq_5_142_detected_df <- pep_nonc %>% 
  filter(protein_accession == "CONTRIB_GENCODE_c5norep142") %>% 
  inner_join(peptide, "peptide") %>% 
  dplyr::count(peptide, start_loc, end_loc) %>% 
  arrange(start_loc) %>% 
  mutate(pep_len = nchar(peptide),
         x_group = c(1,1,1,2,3,3,4,4)) %>% 
  group_by(x_group) %>% 
  mutate(y_order = row_number(),
         y_min = y_order - 0.5 * b_w,
         y_max = y_order + 0.5 * b_w) 

rectangles <- seq_5_142_detected_df %>% 
  group_by(x_group) %>% 
  summarise(min = min(start_loc), max = max(end_loc), .groups = "drop")

seq_5_142_detected_df %>% inner_join(kmer_5_142_p_data, 
                                     c("start_loc" = "kmer_start", "end_loc" = "kmer_end"))

# Create plots
plot_5_142_pred <- ggplot() +
  geom_rect(data = rectangles,
            aes(xmin = min, xmax = max, ymin = 1-b_w, 
                ymax = max(kmer_5_142_p_data$y_max), group = x_group), 
            alpha = 0.2, fill = brewer.pal(5, "Set1")[3]) +
  geom_rect(data = kmer_5_142_p_data,
            aes(xmin = kmer_start, xmax = kmer_end, ymin = y_min, 
                ymax = y_max, fill = min_rank)) +
  black_theme + 
  scale_fill_gradient(name = "Minimum rank", low = grey_col[1], 
                      high = grey_col[7], limits = c(0, 2), 
                      breaks = c(0, 1, 2)) +
  scale_x_continuous(labels = NULL, limits = c(1, 100), expand = c(0,0)) +
  scale_y_continuous(limits = c(1-b_w, NA), expand = c(0,0)) + 
  remove_y +
  theme(axis.line.y = element_blank(), axis.ticks.x = element_blank())

plot_5_142_det <- ggplot() +
  geom_rect(data = rectangles,
            aes(xmin = min, xmax = max, ymin = 1-b_w, 
                ymax = max(seq_5_142_detected_df$y_max), group = x_group), 
            alpha = 0.2, fill = brewer.pal(5, "Set1")[3]) +
  geom_rect(data = seq_5_142_detected_df,
            aes(xmin = start_loc, xmax = end_loc, ymin = y_min, 
                ymax = y_max, fill = n)) +
  black_theme +
  scale_x_continuous(limits = c(1, 100), expand = c(0,0),
                     breaks = 1:100,
                     labels = ~ ifelse(.x %% 10 == 0, .x, ""),
                     name = seq_5_142) +
  scale_fill_gradient(name = "Number of\ndetections\nin MS-runs", 
                      low = brewer.pal(9,"Greens")[6], 
                      high = brewer.pal(9,"Greens")[9], 
                      limits = c(0, NA)) +
  scale_y_reverse(limits = c(NA, 1-b_w)) + 
  remove_y +
  theme(axis.line.y = element_blank())

plot_5_142_shape <- ggplot(seq_5_142_contour) + 
  geom_step(aes(x = amino_acids, y = n), direction = "vh") +
  black_theme +
  scale_x_continuous(limits = c(1, 100), expand = c(0,0), 
                     labels = NULL, name = NULL) +
  scale_y_continuous(name = NULL, expand = c(0,0), breaks = c(0,10,20)) +
  theme(axis.ticks.x = element_blank())

figure_S5 <- plot_5_142_shape / plot_5_142_pred / plot_5_142_det

ggsave("Figure_S5.svg", device = "svg", path = plot_dir, figure_S5, 
       width = 2200, height = 1000, units = "px")
```

# Figures 6 & S5 - Riboseq data

```{r}
ribo_psites <- readRDS(file.path(raw_dir, "ribo_psites.Rds"))
hht_psites <- readRDS(file.path(raw_dir, "hht_psites.Rds"))

ORFs_q <- data.frame(orf_name = c(
    "CONTRIB_GENCODE_c5norep142",
    "CONTRIB_GENCODE_c11riboseqorf4",
    "CONTRIB_GENCODE_c17norep146")) %>% 
  mutate(orf_name = str_replace_all(orf_name, "CONTRIB_GENCODE_", "")) %>% 
  inner_join(orf_sequences, "orf_name") %>% 
  separate_rows(starts,ends) %>% 
  dplyr::rename(start = starts, end = ends) %>% 
  mutate(start = as.integer(start), end = as.integer(end),
         feat = "intORF")

gtf <- fread(gtf_file, skip = 5, col.names = c("chr", "src", "feat", "start", 
                                               "end", "misc1", "strand", 
                                               "misc2", "attributes")) %>% 
  mutate(transcript = str_extract(attributes, "ENST\\d+"))


p_site_colors <- c("#F15A29", brewer.pal(3, "Dark2")[1], "#28296F")


# Create ORF visualizations and save those
sapply(ORFs_q %>% distinct(orf_name) %>% pull(orf_name), function(i) {

  # Get ORF details
  intORF <- ORFs_q %>% filter(orf_name == i) %>% head(1) %>% 
    t() %>% as.data.frame() %>%  rownames_to_column("id") %>% deframe()

  # Get intORF boudaries
  orf_gtf_orf <- ORFs_q %>% 
    filter(orf_name == intORF["orf_name"]) %>% 
    mutate(sort = ifelse(strand == "-", -start, start)) %>% 
    arrange(sort) %>% 
    mutate(y_min = 1.2, y_max = y_min + 1.6,
           color = p_site_colors[1],
           cds_len = end - start + 1,
           cum_len = cumsum(cds_len)-cds_len,
           start_3 = start %% 3,
           len_3 = cum_len %% 3,
           frame = ifelse(strand == "-", ((end %% 3) + len_3) %% 3, 
                          (start_3 - len_3) %% 3)) %>% 
    select(start, end, y_min, y_max, color, frame) %>% 
    arrange(start) 

  # Get full CDS frame + translation frames
  full_trans <- gtf %>% filter(transcript == intORF["transcript"]) %>% 
    filter(feat %in% c("transcript", "CDS", "exon")) %>% 
    mutate(sort = ifelse(strand == "-", -start, start)) %>% 
    arrange(sort) %>% 
    mutate(cds_len = ifelse(feat == "CDS", end - start + 1, 0),
           cum_len = cumsum(cds_len)-cds_len,
           start_3 = start %% 3,
           len_3 = cum_len %% 3,
           frame = ifelse(strand == "-", ((end %% 3) + len_3) %% 3, 
                          (start_3 - len_3) %% 3),
           color = p_site_colors[3],
      y_min = case_when(
      feat == "CDS" ~ -.8,
      feat == "exon" ~ -.5,
      feat == "transcript" ~ -.05),
      y_max = -y_min) 
  
  frame_dif <- orf_gtf_orf %>% 
    rename_with(~paste0(.,"_orf")) %>% 
    cross_join(full_trans) %>% 
    filter(feat == "exon", between(end_orf, start, end)) %>% 
    mutate(diff = (frame_orf - frame) %% 3) %>% 
    head(1) %>% pull(diff)

  # Retrieve global and HHT P-sites
  p_sites_global <- get_psites(ribo_psites, paste0("chr", intORF["chrm"]),
                               min(full_trans$start),
                               max(full_trans$end))
  p_sites_hht <- get_psites(hht_psites, paste0("chr", intORF["chrm"]),
                               min(full_trans$start),
                               max(full_trans$end))
  
  
  
  p_sites_g_exon <- p_sites_global %>% cross_join(full_trans) %>% 
    filter(feat == "exon", between(psite_x, start, end)) %>% 
    mutate(color = case_when(
      frame == group ~ p_site_colors[3],
      (frame + frame_dif) %% 3 == group ~ p_site_colors[1],
      T ~ p_site_colors[2])) %>% select(psite_x, color)
  
  p_sites_g_all <- p_sites_global %>% 
    filter(between(psite_x, min(full_trans$start), max(full_trans$end))) %>% 
    left_join(p_sites_g_exon, "psite_x") %>% 
    replace_na(list(color = grey_col[1]))
  
  p_sites_h_exon <- p_sites_hht %>% cross_join(full_trans) %>% 
    filter(feat == "exon", between(psite_x, start, end)) %>% 
    mutate(color = case_when(
      frame == group ~ p_site_colors[3],
      (frame + frame_dif) %% 3 == group ~ p_site_colors[1],
      T ~ p_site_colors[2])) %>% select(psite_x, color)
  
  p_sites_h_all <- p_sites_hht %>% 
    filter(between(psite_x, min(full_trans$start), max(full_trans$end))) %>% 
    left_join(p_sites_h_exon, "psite_x") %>% 
    replace_na(list(color = grey_col[1]))
  
  
    if (intORF["strand"] == "+") {
      min_zoom = min(orf_gtf_orf$start) - .5
      max_zoom = min_zoom + 42
    } else {
      max_zoom = max(orf_gtf_orf$end) + .5
      min_zoom = max_zoom - 21
    }
  a_site_plot <- ggplot() + 
    geom_bar(data = p_sites_g_all, aes(x = psite_x, y = n, fill = color), 
             stat = "identity", show.legend = F) + black_theme +
    scale_fill_identity() +
      scale_x_continuous(expand = c(0,0),limits = c(min_zoom, 
                                                     max_zoom),
                         labels = c(), name = c())+
      scale_y_continuous(expand = c(0,0), name = "Global P-sites")+
       theme(axis.ticks.x = element_blank())
  
  hht_plot <- ggplot() + 
    geom_bar(data = p_sites_h_all, aes(x = psite_x, y = n, fill = color), 
             stat = "identity", show.legend = F) + black_theme +
    scale_fill_identity() +
      scale_x_continuous(expand = c(0,0),limits = c(min_zoom, 
                                                     max_zoom),
                         labels = c(), name = c())+
      scale_y_continuous(expand = c(0,0), name = "HHT P-sites")+
       theme(axis.ticks.x = element_blank())    
  ribo_plot <- hht_plot / a_site_plot
  # Save plot

  ggsave(paste0("ribo_orf", intORF["strand"], intORF["orf_name"], "_orf_zoom.svg"), 
         device = "svg", path = plot_dir, ribo_plot, width = 1000, 
         height = 1000, units = "px")
  

  return(i)
})
```
